{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f8bd7ced-5e9a-4e8e-a3d4-1e91fc9ff2ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into ‘/usr/local/lib/R/site-library’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"forestploter\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d9eeb993-f367-443d-a155-8ee44bcf4277",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: limma\n",
      "\n",
      "Type 'citation(\"pROC\")' for a citation.\n",
      "\n",
      "\n",
      "Attaching package: ‘pROC’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    cov, smooth, var\n",
      "\n",
      "\n",
      "Loading required package: ggpubr\n",
      "\n",
      "\n",
      "Attaching package: ‘survival’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:survminer’:\n",
      "\n",
      "    myeloma\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(ggplot2)\n",
    "library(pheatmap)\n",
    "library(patchwork)\n",
    "library(edgeR)\n",
    "library(pROC)\n",
    "library(survminer)\n",
    "library(survival)\n",
    "library(forestmodel)\n",
    "library(forestploter)\n",
    "library(grid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9da9b889-210f-4505-b83f-65a1f1069903",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c27d112b-4d3e-45ee-918a-33c472576d80",
   "metadata": {},
   "outputs": [],
   "source": [
    "icb_gsdb <- lapply(read.csv(\"Data/gene_set.csv\"), na.omit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cc1cd815-0f0b-4c3e-bb75-281963d2e8d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "study <- list(\n",
    "    Melanoma_Advanced_GSE91061_Riaz2017 = list(exp = read.csv(\"Data/Melanoma_Advanced_GSE91061_Riaz2017_Log2FPKM.csv\", row.names = 1), pheno = read.csv(\"Data/Melanoma_Advanced_GSE91061_Riaz2017_pheno.csv\")),\n",
    "    Melanoma_Metastatic_GSE78220_Hugo2016 = list(exp = read.csv(\"Data/Melanoma_Metastatic_GSE78220_Hugo2016_Log2FPKM.csv\", row.names = 1), pheno = read.csv(\"Data/Melanoma_Metastatic_GSE78220_Hugo2016_pheno.csv\")),\n",
    "    Melanoma_Metastatic_PRJEB23709_Gide2019 = list(exp = read.csv(\"Data/Melanoma_Metastatic_PRJEB23709_Gide2019_Log2FPKM.csv\", row.names = 1), pheno = read.csv(\"Data/Melanoma_Metastatic_PRJEB23709_Gide2019_pheno.csv\")),\n",
    "    STAD_Metastatic_PRJEB25780_Kim2018 = list(exp = read.csv(\"Data/STAD_Metastatic_PRJEB25780_Kim2018_Log2FPKM.csv\", row.names = 1), pheno = read.csv(\"Data/STAD_Metastatic_PRJEB25780_Kim2018_pheno.csv\")),\n",
    "    NSCLC_Advanced_GSE135222_Kim2020 = list(exp = read.csv(\"Data/NSCLC_Advanced_GSE135222_Kim2020_Log2TPM.csv\", row.names = 1), pheno = read.csv(\"Data/NSCLC_Advanced_GSE135222_Kim2020_pheno.csv\")),\n",
    "    NSCLC_Metastatic_GSE136961_Hwang2020 = list(exp = read.csv(\"Data/NSCLC_Metastatic_GSE136961_Hwang2020_Log2TPM.csv\", row.names = 1), pheno = read.csv(\"Data/NSCLC_Metastatic_GSE136961_Hwang2020_pheno.csv\")),\n",
    "    SU2C = list(exp = read.csv(\"Data/SU2C_Log2TPM.csv\", row.names = 1), pheno = read.csv(\"Data/SU2C_pheno.csv\"))\n",
    ")\n",
    "\n",
    "for(pname in names(study)){\n",
    "    rownames(study[[pname]]$pheno) <- study[[pname]]$pheno$sid\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e625e19c-ee57-45f2-971b-93cf89ad62fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b8e1a7ef-99e7-496c-8238-8728af04881d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Melanoma_Advanced_GSE91061_Riaz2017 \n",
      "Melanoma_Metastatic_GSE78220_Hugo2016 \n",
      "Melanoma_Metastatic_PRJEB23709_Gide2019 \n",
      "STAD_Metastatic_PRJEB25780_Kim2018 \n",
      "NSCLC_Advanced_GSE135222_Kim2020 \n",
      "NSCLC_Metastatic_GSE136961_Hwang2020 \n",
      "SU2C \n"
     ]
    }
   ],
   "source": [
    "# compute z-scores for gene sets\n",
    "\n",
    "for(pname in names(study)){\n",
    "    cat(pname,\"\\n\")\n",
    "    \n",
    "    ngene_ovlp <- sapply(icb_gsdb, function(x){length(intersect(rownames(study[[pname]]$exp), x))})\n",
    "    valid_gs <- names(ngene_ovlp)[ngene_ovlp > 0]\n",
    "    \n",
    "    temp <- data.frame(matrix(NA, ncol=ncol(study[[pname]]$exp), nrow=length(valid_gs)))\n",
    "    rownames(temp) <- valid_gs\n",
    "    colnames(temp) <- colnames(study[[pname]]$exp)\n",
    "\n",
    "    for(the_gs in rownames(temp)){\n",
    "        genes <- intersect(icb_gsdb[[the_gs]], rownames(study[[pname]]$exp))\n",
    "        if(length(genes) == 1){\n",
    "            temp[the_gs, ] <- as.numeric(scale(as.numeric(study[[pname]]$exp[genes,])))\n",
    "        } else {\n",
    "            temp[the_gs, ] <- rowMeans(scale(t(study[[pname]]$exp[genes,])))\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    if(any(is.na(temp))){\n",
    "        temp[is.na(temp)] <- 0\n",
    "    }\n",
    "    \n",
    "    if(any(apply(temp == 0, 1, all))){\n",
    "        temp <- temp[-which(apply(temp == 0, 1, all)),]\n",
    "    }\n",
    "    \n",
    "    study[[pname]]$gs_zscore <- temp\n",
    "    \n",
    "    study[[pname]]$pheno <- cbind(study[[pname]]$pheno, t(temp))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "15e28370-52f4-4773-9e61-1371c927811f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Melanoma_Advanced_GSE91061_Riaz2017 : 169 gene sets.\n",
      "Melanoma_Metastatic_GSE78220_Hugo2016 : 168 gene sets.\n",
      "Melanoma_Metastatic_PRJEB23709_Gide2019 : 167 gene sets.\n",
      "STAD_Metastatic_PRJEB25780_Kim2018 : 169 gene sets.\n",
      "NSCLC_Advanced_GSE135222_Kim2020 : 169 gene sets.\n",
      "NSCLC_Metastatic_GSE136961_Hwang2020 : 130 gene sets.\n",
      "SU2C : 168 gene sets.\n"
     ]
    }
   ],
   "source": [
    "for(pname in names(study)){\n",
    "    cat(pname,\":\",nrow(study[[pname]]$gs_zscore),\"gene sets.\\n\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b45f496-2212-4d80-8c6c-5b7dc343887f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "64641456-f9af-4ced-aee9-eae0980a4aca",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Median gene set size for constructing the random gene set: 20 \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# compute null distribution of AUCs for the combined model by combining T_Inflamed with random gene sets\n",
    "\n",
    "ntop <- median(sapply(icb_gsdb, length))\n",
    "cat(\"Median gene set size for constructing the random gene set:\", ntop,\"\\n\")\n",
    "\n",
    "for(one_study in names(study)){\n",
    "    study[[one_study]]$rand_model <- list()\n",
    "\n",
    "    set.seed(1357)\n",
    "    study[[one_study]]$rand_model$pheno_rand <- data.frame(\n",
    "        Response_Bin = study[[one_study]]$pheno$Response_Bin,\n",
    "        T_Inflamed = as.numeric(study[[one_study]]$gs_zscore[\"T_Inflamed\",]),\n",
    "        randz=0, randg=0\n",
    "    )\n",
    "    \n",
    "    med_gsize <- ntop\n",
    "    study[[one_study]]$rand_model[[one_study]]$auc_randz <- c()\n",
    "    study[[one_study]]$rand_model[[one_study]]$auc_randg <- c()\n",
    "    \n",
    "    # Null AUC is constructed by 100 random combined models\n",
    "    # Random combined models are constructed in two ways: random z-scores and random gene sets\n",
    "    for(i in 1:100){\n",
    "        study[[one_study]]$rand_model$pheno_rand$randz <- rnorm(nrow(study[[one_study]]$rand_model$pheno_rand))\n",
    "        ind_gene <- sample(1:nrow(study[[one_study]]$exp),med_gsize)\n",
    "        study[[one_study]]$rand_model$pheno_rand$randg <- rowMeans(scale(t(study[[one_study]]$exp[ind_gene,])))\n",
    "\n",
    "        test <- summary(glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed + randz\")), family = \"binomial\", data = study[[one_study]]$rand_model$pheno_rand))\n",
    "        pred_randz <- test$coefficients[\"T_Inflamed\",1]*study[[one_study]]$rand_model$pheno_rand$T_Inflamed+test$coefficients[\"randz\",1]*study[[one_study]]$rand_model$pheno_rand$randz\n",
    "        study[[one_study]]$rand_model$auc_randz <- c(study[[one_study]]$rand_model$auc_randz, auc(study[[one_study]]$rand_model$pheno_rand$Response_Bin, pred_randz, direction=\"<\"))\n",
    "\n",
    "        test <- summary(glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed + randg\")), family = \"binomial\", data = study[[one_study]]$rand_model$pheno_rand))\n",
    "        pred_randg <- test$coefficients[\"T_Inflamed\",1]*study[[one_study]]$rand_model$pheno_rand$T_Inflamed+test$coefficients[\"randg\",1]*study[[one_study]]$rand_model$pheno_rand$randg\n",
    "        study[[one_study]]$rand_model$auc_randg <- c(study[[one_study]]$rand_model$auc_randg, auc(study[[one_study]]$rand_model$pheno_rand$Response_Bin, pred_randg, direction=\"<\"))\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2753f25e-e985-470b-9cf1-aef749b850cc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "55e643c4-6416-4f5a-a33a-4575afb9ab73",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Melanoma_Advanced_GSE91061_Riaz2017\n",
      "\n",
      "Melanoma_Metastatic_GSE78220_Hugo2016\n",
      "\n",
      "Melanoma_Metastatic_PRJEB23709_Gide2019\n",
      "\n",
      "STAD_Metastatic_PRJEB25780_Kim2018\n",
      "\n",
      "NSCLC_Advanced_GSE135222_Kim2020\n",
      "\n",
      "NSCLC_Metastatic_GSE136961_Hwang2020\n",
      "\n",
      "SU2C\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Predict response using individual signatures\n",
    "\n",
    "for(pname in names(study)){\n",
    "    message(pname)\n",
    "\n",
    "    res_temp <- data.frame(matrix(NA, nrow=nrow(study[[pname]]$gs_zscore), ncol=4))\n",
    "    colnames(res_temp) <- c(\"beta\",\"p\",\"AIC\",\"AUC\")\n",
    "    rownames(res_temp) <- rownames(study[[pname]]$gs_zscore)\n",
    "\n",
    "    res_temp$beta <- 0\n",
    "    res_temp$p <- 1\n",
    "\n",
    "    for(the_gs in rownames(res_temp)){\n",
    "\n",
    "        test <- summary(glm(as.formula(paste0(\"Response_Bin ~ \",the_gs)), family = \"binomial\", data = study[[pname]]$pheno))\n",
    "\n",
    "        if(paste0(the_gs) %in% rownames(test$coefficients)){\n",
    "            test_ind <- which(paste0(the_gs) == rownames(test$coefficients))\n",
    "            res_temp[the_gs, \"beta\"] <- test$coefficients[test_ind,1]\n",
    "            res_temp[the_gs, \"p\"] <- test$coefficients[test_ind,4]\n",
    "        }\n",
    "        res_temp[the_gs, \"AIC\"] <- test$aic\n",
    "\n",
    "        pred <- ROCR::prediction(study[[pname]]$pheno[, paste0(the_gs)]*res_temp[the_gs,\"beta\"], study[[pname]]$pheno[, \"Response_Bin\"])\n",
    "        res_temp[the_gs, \"AUC\"] <- ROCR::performance(pred, measure = \"auc\")@y.values[[1]]\n",
    "    }\n",
    "    \n",
    "    study[[pname]]$uni_gs <- res_temp\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fcb7acf9-94ef-4861-81b7-bc3bc7223204",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "331bf7df-7f2d-46e4-acac-ad8d0a8ec462",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure 1A: Association between response and individual signatures\n",
    "\n",
    "pos_sig <- study$SU2C$uni_gs[study$SU2C$uni_gs$beta > 0,]\n",
    "neg_sig <- study$SU2C$uni_gs[study$SU2C$uni_gs$beta < 0,]\n",
    "pos_sig <- pos_sig[order(pos_sig$p),]\n",
    "neg_sig <- neg_sig[order(neg_sig$p),]\n",
    "\n",
    "ntop <- 10\n",
    "\n",
    "dat.plot <- rbind(pos_sig[1:ntop,], neg_sig[1:ntop,])\n",
    "dat.plot$SignedLog10P <- -log10(dat.plot$p)*sign(dat.plot$beta)\n",
    "dat.plot$GeneSet <- rownames(dat.plot)\n",
    "dat.plot$GeneSet <- factor(dat.plot$GeneSet, levels = rev(dat.plot$GeneSet))\n",
    "dat.plot$Association <- c(rep(\"CR/PR\",ntop), rep(\"PD/SD\",ntop))\n",
    "\n",
    "p <- ggplot(data = dat.plot, aes(x=GeneSet, y=SignedLog10P, fill=Association)) + geom_bar(stat = \"identity\") +\n",
    "    geom_hline(yintercept = c(log10(0.05), -log10(0.05)), linetype=\"dashed\", color=\"blue\") +\n",
    "    coord_flip() + theme_bw() + theme(legend.position = \"top\") + scale_fill_brewer(palette = \"Set1\")\n",
    "ggsave(\"Figure/Figure1A.pdf\", p, width=6, height=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "245e7d22-1478-4f09-8cc9-5ed81a7fe07f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "38cd408b-68a7-45bf-b6e1-dd19b09b63ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Predict response by combining T_Inflamed with other signatures\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "gs_vec <- rownames(study[[pname]]$gs_zscore)\n",
    "gs_vec <- gs_vec[gs_vec != \"T_Inflamed\"]\n",
    "\n",
    "res_temp <- data.frame(matrix(NA, nrow=length(gs_vec), ncol=10))\n",
    "colnames(res_temp) <- c(\"v1\",\"v2\",\"b1\",\"b2\",\"p1\",\"p2\",\"AIC1\",\"AIC2\",\"AIC\",\"AUC\")\n",
    "\n",
    "res_temp$b1 <- 0\n",
    "res_temp$p1 <- 1\n",
    "res_temp$b2 <- 0\n",
    "res_temp$p2 <- 1\n",
    "\n",
    "ind <- 1\n",
    "for(i in 1:length(gs_vec)){\n",
    "    v1 <- gs_vec[i]\n",
    "    v2 <- \"T_Inflamed\"\n",
    "\n",
    "    ### compute zscore\n",
    "\n",
    "    test <- summary(glm(as.formula(paste0(\"Response_Bin ~ \",v1,\" + \",v2)), family = \"binomial\", data = study[[pname]]$pheno))\n",
    "\n",
    "    res_temp[ind, \"v1\"] <- v1\n",
    "    if(v1 %in% rownames(test$coefficients)){\n",
    "        test_ind <- which(v1 == rownames(test$coefficients))\n",
    "        res_temp[ind, \"b1\"] <- test$coefficients[test_ind,1]\n",
    "        res_temp[ind, \"p1\"] <- test$coefficients[test_ind,4]\n",
    "    }\n",
    "    res_temp[ind, \"v2\"] <- v2\n",
    "    if(v2 %in% rownames(test$coefficients)){\n",
    "        test_ind <- which(v2 == rownames(test$coefficients))\n",
    "        res_temp[ind, \"b2\"] <- test$coefficients[test_ind,1]\n",
    "        res_temp[ind, \"p2\"] <- test$coefficients[test_ind,4]\n",
    "    }\n",
    "    res_temp[ind, \"AIC\"] <- test$aic\n",
    "    res_temp[ind, \"AIC1\"] <- test$aic-study[[pname]]$uni_gs[v1, \"AIC\"]\n",
    "    res_temp[ind, \"AIC2\"] <- test$aic-study[[pname]]$uni_gs[v2, \"AIC\"]\n",
    "\n",
    "    temp_pred <- study[[pname]]$pheno[, v1]*res_temp[ind, \"b1\"]+study[[pname]]$pheno[, v2]*res_temp[ind, \"b2\"]\n",
    "    pred <- ROCR::prediction(temp_pred, study[[pname]]$pheno[, \"Response_Bin\"])\n",
    "    res_temp[ind, \"AUC\"] <- ROCR::performance(pred, measure = \"auc\")@y.values[[1]]\n",
    "\n",
    "    ind <- ind+1\n",
    "}\n",
    "\n",
    "study[[pname]]$pair_gs <- res_temp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a59bbd1-9493-4d78-9365-628abfa40fde",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "163047eb-9d53-4984-9404-2ec257927905",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da68c392-5a22-4f12-8eec-5cf62aeb5d36",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c94292c4-9ac7-4954-b8c4-e23bd5d63ca6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure 1B: P-value changes from univariate models to combined models (with T_Inflamed) for top 10 negative signatures\n",
    "\n",
    "pos_sig <- study$SU2C$uni_gs[study$SU2C$uni_gs$beta > 0,]\n",
    "neg_sig <- study$SU2C$uni_gs[study$SU2C$uni_gs$beta < 0,]\n",
    "pos_sig <- pos_sig[order(pos_sig$p),]\n",
    "neg_sig <- neg_sig[order(neg_sig$p),]\n",
    "\n",
    "ntop <- 10\n",
    "\n",
    "selected_gs <- c(rownames(neg_sig)[1:ntop])\n",
    "pivot_gs <- \"T_Inflamed\"\n",
    "\n",
    "gep_pair <- study$SU2C$pair_gs[(study$SU2C$pair_gs$v1 %in% selected_gs & study$SU2C$pair_gs$v2 == \"T_Inflamed\"),]\n",
    "gep_pair <- gep_pair[match(selected_gs, gep_pair$v1), ]\n",
    "\n",
    "dat.plot <- data.frame(matrix(NA, ncol=5, nrow=4*ntop))\n",
    "colnames(dat.plot) <- c(\"pval\",\"beta\",\"Model\",\"Group\",\"GeneSet\")\n",
    "\n",
    "dat.plot[1:ntop, \"pval\"] <- study$SU2C$uni_gs[pivot_gs,\"p\"]\n",
    "dat.plot[1:ntop, \"beta\"] <- study$SU2C$uni_gs[pivot_gs,\"beta\"]\n",
    "dat.plot[1:ntop, \"Model\"] <- \"Univariate\"\n",
    "dat.plot[1:ntop, \"Group\"] <- paste(pivot_gs, \":\", selected_gs)\n",
    "\n",
    "dat.plot[(ntop+1):(2*ntop), \"pval\"] <- study$SU2C$uni_gs[selected_gs,\"p\"]\n",
    "dat.plot[(ntop+1):(2*ntop), \"beta\"] <- study$SU2C$uni_gs[selected_gs,\"beta\"]\n",
    "dat.plot[(ntop+1):(2*ntop), \"Model\"] <- \"Univariate\"\n",
    "dat.plot[(ntop+1):(2*ntop), \"Group\"] <- selected_gs\n",
    "\n",
    "dat.plot[(2*ntop+1):(4*ntop), \"Model\"] <- \"Combined\"\n",
    "\n",
    "for(i in 1:nrow(gep_pair)){\n",
    "    pivot_p <- gep_pair[i, \"p1\"]\n",
    "    pivot_b <- gep_pair[i, \"b1\"]\n",
    "    if(gep_pair[i,\"v2\"] == pivot_gs){\n",
    "        pivot_p <- gep_pair[i, \"p2\"]\n",
    "        pivot_b <- gep_pair[i, \"b2\"]\n",
    "    }\n",
    "    \n",
    "    target_p <- gep_pair[i, \"p2\"]\n",
    "    target_b <- gep_pair[i, \"b2\"]\n",
    "    target_gs <- gep_pair[i, \"v2\"]\n",
    "    if(gep_pair[i,\"v1\"] != pivot_gs){\n",
    "        target_p <- gep_pair[i, \"p1\"]\n",
    "        target_b <- gep_pair[i, \"b1\"]\n",
    "        target_gs <- gep_pair[i, \"v1\"]\n",
    "    }\n",
    "    \n",
    "    dat.plot[2*ntop+i, \"pval\"] <- pivot_p\n",
    "    dat.plot[2*ntop+i, \"beta\"] <- pivot_b\n",
    "    dat.plot[2*ntop+i, \"Group\"] <- paste(pivot_gs, \":\", target_gs)\n",
    "    \n",
    "    dat.plot[3*ntop+i, \"pval\"] <- target_p\n",
    "    dat.plot[3*ntop+i, \"beta\"] <- target_b\n",
    "    dat.plot[3*ntop+i, \"Group\"] <- paste(target_gs)\n",
    "    \n",
    "    dat.plot[i, \"GeneSet\"] <- pivot_gs\n",
    "    dat.plot[ntop + i, \"GeneSet\"] <- \"Top10_Negative\"\n",
    "    dat.plot[2*ntop + i, \"GeneSet\"] <- pivot_gs\n",
    "    dat.plot[3*ntop + i, \"GeneSet\"] <- \"Top10_Negative\"\n",
    "}\n",
    "\n",
    "dat.plot$Model <- factor(dat.plot$Model, levels=c(\"Univariate\",\"Combined\"))\n",
    "dat.plot$SignedLog10P <- -log10(dat.plot$pval)*sign(dat.plot$beta)\n",
    "p <- ggplot(data = dat.plot, aes(x=Model, y=SignedLog10P, group=Group)) + geom_line() + geom_point(aes(color=GeneSet), size=2) +\n",
    "    geom_hline(yintercept = c(-log10(0.05), 0, log10(0.05)), linetype=c(\"dashed\",\"solid\",\"dashed\"), color=c(\"blue\",\"darkgrey\",\"blue\")) +\n",
    "    theme_bw() + scale_color_brewer(palette = \"Set1\")\n",
    "ggsave(\"Figure/Figure1B.pdf\", p, width=6, height=4)\n",
    "\n",
    "\n",
    "\n",
    "# # Figure 1C: AIC changes from univariate models to combined models (with T_Inflamed) for top 10 negative signatures\n",
    "\n",
    "temp_t <- study$SU2C$pair_gs[study$SU2C$pair_gs$v2 == \"T_Inflamed\",]\n",
    "dat.plot <- data.frame(\n",
    "    Model = c(rep(\"Univariate\", 11), rep(\"Combined with\\nT_Inflamed\", 10)),\n",
    "    AIC = c(study$SU2C$uni_gs[c(\"T_Inflamed\", selected_gs),\"AIC\"], temp_t[match(selected_gs, temp_t$v1), \"AIC\"]),\n",
    "    GeneSet = c(\"T_Inflamed\",selected_gs, selected_gs)\n",
    ")\n",
    "\n",
    "dat.plot$Model <- factor(dat.plot$Model, levels=c(\"Univariate\",\"Combined with\\nT_Inflamed\"))\n",
    "\n",
    "p <- ggplot(data = dat.plot, aes(x=Model, y=AIC, group=GeneSet, color=GeneSet)) + geom_point() + geom_line() + theme_bw()\n",
    "ggsave(\"Figure/Figure1C.pdf\", p, width=6, height=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ebb2488c-33b5-47ab-9700-bb6a93ce2bae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure 1D: AUC changes from univariate models to combined models (with T_Inflamed) for top 10 negative signatures\n",
    "\n",
    "pos_sig <- study$SU2C$uni_gs[study$SU2C$uni_gs$beta > 0,]\n",
    "neg_sig <- study$SU2C$uni_gs[study$SU2C$uni_gs$beta < 0,]\n",
    "pos_sig <- pos_sig[order(pos_sig$p),]\n",
    "neg_sig <- neg_sig[order(neg_sig$p),]\n",
    "\n",
    "ntop <- 10\n",
    "\n",
    "selected_gs <- c(rownames(neg_sig)[1:ntop])\n",
    "\n",
    "dat.plot <- data.frame(\n",
    "    GeneSet = c(paste0(\"T_Inflamed+\", selected_gs), \"T_Inflamed+randg at p=0.05\", \"T_Inflamed\"),\n",
    "    AUC = 0,\n",
    "    Type = c(rep(\"Real\", length(selected_gs)), \"Random\",\"Real\")\n",
    ")\n",
    "rownames(dat.plot) <- c(selected_gs, \"randg\", \"T_Inflamed\")\n",
    "dat.plot$GeneSet <- factor(dat.plot$GeneSet, levels = dat.plot$GeneSet)\n",
    "\n",
    "dat.plot$AUC[which(dat.plot$GeneSet == \"T_Inflamed\")] <- study$SU2C$uni_gs[\"T_Inflamed\", \"AUC\"]\n",
    "dat.plot$AUC[which(dat.plot$GeneSet == \"T_Inflamed+randg at p=0.05\")] <- sort(study$SU2C$rand_model$auc_randg)[95]\n",
    "\n",
    "dat.plot[selected_gs, \"AUC\"] <- study$SU2C$pair_gs[match(selected_gs, study$SU2C$pair_gs$v1), \"AUC\"]\n",
    "\n",
    "\n",
    "p <- ggplot(data = dat.plot, aes(x=GeneSet, y=AUC, fill=Type)) + geom_bar(stat=\"identity\") +\n",
    "    geom_hline(yintercept = c(dat.plot$AUC[nrow(dat.plot)], dat.plot[\"randg\",\"AUC\"]), linetype=\"dotted\", color=c(\"blue\",\"red\")) +\n",
    "    scale_fill_manual(values = c(\"grey\",\"#505050\")) + \n",
    "    coord_flip(ylim=c(0.5,0.7)) + theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1), legend.position = \"none\")\n",
    "ggsave(\"Figure/Figure1D.pdf\", p, width=6, height=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad962170-ab7e-4878-8bde-1c2b8ac9ebcc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c7dcd2aa-bf0d-4c14-a1d4-2a97af61156a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select myeloid-based signatures that significantly better predicted response when combining with T_Inflamed\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "pair_filter <- study[[pname]]$pair_gs$v2 == \"T_Inflamed\" & study[[pname]]$pair_gs$b1 < 0 & study[[pname]]$pair_gs$b2 > 0 &\n",
    "    study[[pname]]$pair_gs$p1 < 0.05 & study[[pname]]$pair_gs$p2 < 0.05 &\n",
    "    study[[pname]]$pair_gs$p1 < study[[pname]]$uni_gs[study[[pname]]$pair_gs$v1,\"p\"] &\n",
    "    study[[pname]]$pair_gs$p2 < study[[pname]]$uni_gs[study[[pname]]$pair_gs$v2,\"p\"] &\n",
    "    study[[pname]]$pair_gs$AUC > sort(study[[pname]]$rand_model$auc_randg)[95]\n",
    "\n",
    "study[[pname]]$pair_neg <- list()\n",
    "study[[pname]]$pair_neg$stat <- study[[pname]]$pair_gs[pair_filter,]\n",
    "study[[pname]]$pair_neg$gs <- study[[pname]]$pair_gs[pair_filter, \"v1\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd0ba0e0-b9f0-47e1-96a5-f60962db3517",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c634889d-6b79-4237-8af6-9ee2a2f0035b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SFigure 1\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "dat.plot <- data.frame(matrix(0, nrow = length(study[[pname]]$pair_neg$gs), ncol=2))\n",
    "rownames(dat.plot) <- study[[pname]]$pair_neg$gs\n",
    "colnames(dat.plot) <- c(\"T_Inflamed\",\"Myeloid\")\n",
    "\n",
    "lr.t <- glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed\")), family = \"binomial\", data = study[[pname]]$pheno)\n",
    "for(one_gs in study[[pname]]$pair_neg$gs){\n",
    "    lr.uni <- glm(as.formula(paste0(\"Response_Bin ~ \", one_gs)), family = \"binomial\", data = study[[pname]]$pheno)\n",
    "    lr.comb <- glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed + \", one_gs)), family = \"binomial\", data = study[[pname]]$pheno)\n",
    "    dat.plot[one_gs, \"T_Inflamed\"] <- anova(lr.t,lr.comb,test=\"LRT\")[2,\"Pr(>Chi)\"]\n",
    "    dat.plot[one_gs, \"Myeloid\"] <- anova(lr.uni,lr.comb,test=\"LRT\")[2,\"Pr(>Chi)\"]\n",
    "}\n",
    "\n",
    "dat.plot <- -log10(dat.plot)\n",
    "dat.plot$Model <- paste0(\"Response ~ T_Inflamed + \", study[[pname]]$pair_neg$gs)\n",
    "\n",
    "\n",
    "p1 <- ggplot(data = dat.plot, aes(x = Model, y = T_Inflamed)) + geom_bar(stat = \"identity\") + geom_hline(yintercept = -log10(0.05), color = \"blue\", linetype=\"dashed\") +\n",
    "    labs(title = \"Response ~ T_Inflamed\", y = \"-Log10P (LRT test)\") + coord_flip() + theme_bw()\n",
    "p2 <- ggplot(data = dat.plot, aes(x = Model, y = Myeloid)) + geom_bar(stat = \"identity\") + geom_hline(yintercept = -log10(0.05), color = \"blue\", linetype=\"dashed\") +\n",
    "    labs(title = \"Response ~ NegSignature\", y = \"-Log10P (LRT test)\") + coord_flip() + theme_bw() + theme(axis.text.y = element_blank(), axis.title.y = element_blank())\n",
    "ggsave(\"Figure/SFigure1.pdf\", p1+p2, width = 10, height = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "124fb1d7-d0d2-4580-b5c0-cfb15707861e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "52bd9715-c44e-4b25-b9ad-0f98f35a6516",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Load NSCLC scRNA-seq data from Salcher et al. (2022)\n",
    "\n",
    "library(Seurat)\n",
    "\n",
    "sobj <- readRDS(\"~/datasets/scRNA/references/lung/Neutrophils_NSCLC_Salcher2022/core_symbol.rds\")\n",
    "sobj <- subset(sobj, cells = colnames(sobj)[sobj@meta.data$disease %in% c(\"non-small cell lung carcinoma\",\n",
    "                                                                          \"squamous cell lung carcinoma\",\"lung adenocarcinoma\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6bbd5846-c914-4065-847e-e159a98003e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure 2: expression of myeloid-based signatures across NSCLC cell types\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "show_gs <- c()\n",
    "\n",
    "for(one_gs in c(\"T_Inflamed\", study[[pname]]$pair_neg$gs)){\n",
    "    valid_genes <- intersect(rownames(sobj[[\"RNA\"]]), icb_gsdb[[one_gs]])\n",
    "    if(length(valid_genes) == 0){\n",
    "        next\n",
    "    }\n",
    "\n",
    "    if(length(valid_genes) == 1){\n",
    "        sobj@meta.data[[one_gs]] <- as.numeric(sobj@assays$RNA@data[valid_genes,])\n",
    "    } else {\n",
    "        sobj@meta.data[[one_gs]] <- colMeans(sobj@assays$RNA@data[valid_genes,])\n",
    "    }\n",
    "\n",
    "    show_gs <- c(show_gs, one_gs)\n",
    "}\n",
    "\n",
    "gs_mtx <- data.frame(matrix(NA, nrow=length(unique(sobj@meta.data$cell_type)), ncol=length(show_gs)))\n",
    "rownames(gs_mtx) <- unique(sobj@meta.data$cell_type)\n",
    "colnames(gs_mtx) <- show_gs\n",
    "\n",
    "for(ctype in rownames(gs_mtx)){\n",
    "    gs_mtx[ctype, ] <- colMeans(sobj@meta.data[sobj@meta.data$cell_type == ctype, show_gs])\n",
    "}\n",
    "\n",
    "col.anno <- data.frame(\n",
    "    nLog10P = c(NA, -log10(study[[pname]]$pair_neg$stat[match(show_gs[2:length(show_gs)], study[[pname]]$pair_neg$stat$v1), \"p1\"])),\n",
    "    AUC = c(NA, study[[pname]]$pair_neg$stat[match(show_gs[2:length(show_gs)], study[[pname]]$pair_neg$stat$v1), \"AUC\"])\n",
    ")\n",
    "rownames(col.anno) <- show_gs\n",
    "\n",
    "temp <- scale(gs_mtx)\n",
    "temp[temp < -2] <- -2\n",
    "temp[temp > 2] <- 2\n",
    "\n",
    "pheatmap::pheatmap(temp, breaks = seq(-2,2,4/100), color = colorRampPalette(rev(RColorBrewer::brewer.pal(n = 7, name = \"RdBu\")))(100),\n",
    "         filename=\"Figure/Figure2.pdf\", width=11, height=9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d03ac9d-76df-4e34-b8d1-4e31d4c778a6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dd9dd37f-461c-495f-9212-4248d9704acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SFigure 3: Deconvolution TIMER\n",
    "\n",
    "deconv_est <- read.csv(\"Data/TIMER_est.csv\")\n",
    "rownames(deconv_est) <- deconv_est[,1]\n",
    "deconv_est <- deconv_est[,-1]\n",
    "\n",
    "deconv_est <- deconv_est[, rownames(study$SU2C$pheno)]\n",
    "\n",
    "cor_mat <- data.frame(matrix(0, nrow = length(study$SU2C$pair_neg$gs), ncol = nrow(deconv_est)))\n",
    "rownames(cor_mat) <- study$SU2C$pair_neg$gs\n",
    "colnames(cor_mat) <- rownames(deconv_est)\n",
    "\n",
    "for(one_gs in rownames(cor_mat)){\n",
    "    for(one_ct in colnames(cor_mat)){\n",
    "        cor_mat[one_gs, one_ct] <- cor(study$SU2C$pheno[[one_gs]], as.numeric(deconv_est[one_ct,]))\n",
    "    }\n",
    "}\n",
    "\n",
    "pheatmap::pheatmap(cor_mat, breaks = seq(-0.5,0.5,1/100), color = colorRampPalette(rev(RColorBrewer::brewer.pal(n = 7, name = \"RdBu\")))(100),\n",
    "         filename=\"Figure/SFigure3_TIMER.pdf\", width=6, height=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "bbadb46e-3201-4b26-beaa-7d059e31497e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SFigure 3: Deconvolution iSort\n",
    "\n",
    "deconv_est <- read.csv(\"Data/iSort_out.csv\", row.names=1)\n",
    "\n",
    "deconv_est <- deconv_est[rownames(study$SU2C$pheno),]\n",
    "\n",
    "cor_mat <- data.frame(matrix(0, nrow = length(study$SU2C$pair_neg$gs), ncol = ncol(deconv_est)))\n",
    "rownames(cor_mat) <- study$SU2C$pair_neg$gs\n",
    "colnames(cor_mat) <- colnames(deconv_est)\n",
    "\n",
    "for(one_gs in rownames(cor_mat)){\n",
    "    for(one_ct in colnames(cor_mat)){\n",
    "        cor_mat[one_gs, one_ct] <- cor(study$SU2C$pheno[[one_gs]], as.numeric(deconv_est[,one_ct]))\n",
    "    }\n",
    "}\n",
    "\n",
    "pheatmap::pheatmap(cor_mat, breaks = seq(-0.5,0.5,1/100), color = colorRampPalette(rev(RColorBrewer::brewer.pal(n = 7, name = \"RdBu\")))(100),\n",
    "         filename=\"Figure/SFigure3_iSort.pdf\", width=10, height=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26e092d1-e217-41b3-a493-631cd218d847",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a6aa8136-ad87-4c75-9aed-17bf3290ac69",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Cross-validation\n",
    "\n",
    "frac <- 0.5\n",
    "\n",
    "better_auc <- data.frame(matrix(0, nrow = length(study$SU2C$pair_neg$gs), ncol = 2))\n",
    "rownames(better_auc) <- study$SU2C$pair_neg$gs\n",
    "colnames(better_auc) <- c(\"T_Inflamed\",\"NegSignature\")\n",
    "\n",
    "set.seed(1357)\n",
    "\n",
    "for(i in 1:100){\n",
    "\n",
    "    train_ind <- sample(1:nrow(study$SU2C$pheno), as.integer(nrow(study$SU2C$pheno)*frac))\n",
    "    test_ind <- setdiff(1:nrow(study$SU2C$pheno), train_ind)\n",
    "\n",
    "    for(one_gs in study$SU2C$pair_neg$gs){\n",
    "        fit.t <- summary(glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed\")), family = \"binomial\", data = study[[pname]]$pheno[train_ind,]))\n",
    "        fit.n <- summary(glm(as.formula(paste0(\"Response_Bin ~ \", one_gs)), family = \"binomial\", data = study[[pname]]$pheno[train_ind,]))\n",
    "        fit.comb <- summary(glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed + \", one_gs)), family = \"binomial\", data = study[[pname]]$pheno[train_ind,]))\n",
    "\n",
    "        b_t_alone <- fit.t$coefficients[2,1]\n",
    "        b_n_alone <- fit.n$coefficients[2,1]\n",
    "        b_t_comb <- fit.comb$coefficients[2,1]\n",
    "        b_n_comb <- fit.comb$coefficients[3,1]\n",
    "        \n",
    "#         auc_t_alone <- auc(study$SU2C$pheno[test_ind, \"Response_Bin\"], study$SU2C$pheno[[\"T_Inflamed\"]][test_ind]*b_t_alone, direction=\"<\")\n",
    "#         auc_n_alone <- auc(study$SU2C$pheno[test_ind, \"Response_Bin\"], study$SU2C$pheno[[one_gs]][test_ind]*b_n_alone, direction=\"<\")\n",
    "#         auc_comb <- auc(study$SU2C$pheno[test_ind, \"Response_Bin\"], study$SU2C$pheno[[one_gs]][test_ind]*b_n_comb+study$SU2C$pheno[[\"T_Inflamed\"]][test_ind]*b_t_comb, direction=\"<\")\n",
    "\n",
    "        pred <- ROCR::prediction(study$SU2C$pheno[[\"T_Inflamed\"]][test_ind]*b_t_alone, study$SU2C$pheno[test_ind, \"Response_Bin\"])\n",
    "        auc_t_alone <- ROCR::performance(pred, measure = \"auc\")@y.values[[1]]\n",
    "        pred <- ROCR::prediction(study$SU2C$pheno[[one_gs]][test_ind]*b_n_alone, study$SU2C$pheno[test_ind, \"Response_Bin\"])\n",
    "        auc_n_alone <- ROCR::performance(pred, measure = \"auc\")@y.values[[1]]\n",
    "        pred <- ROCR::prediction(study$SU2C$pheno[[one_gs]][test_ind]*b_n_comb+study$SU2C$pheno[[\"T_Inflamed\"]][test_ind]*b_t_comb, study$SU2C$pheno[test_ind, \"Response_Bin\"])\n",
    "        auc_comb <- ROCR::performance(pred, measure = \"auc\")@y.values[[1]]\n",
    "\n",
    "        if(auc_comb > auc_t_alone) better_auc[one_gs, \"T_Inflamed\"] <- better_auc[one_gs, \"T_Inflamed\"]+1\n",
    "        if(auc_comb > auc_n_alone) better_auc[one_gs, \"NegSignature\"] <- better_auc[one_gs, \"NegSignature\"]+1\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "942b5ee4-bbac-41e6-886e-cc131abf4a0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "23"
      ],
      "text/latex": [
       "23"
      ],
      "text/markdown": [
       "23"
      ],
      "text/plain": [
       "[1] 23"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Least significant p-value: 9.050013e-08 \n"
     ]
    }
   ],
   "source": [
    "sum(apply(better_auc >= 75, 1, all))\n",
    "cat(\"Least significant p-value:\",pbinom(q = min(better_auc[,c(\"T_Inflamed\",\"NegSignature\")]), size = 100, prob = 0.5, lower.tail = F),\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "31a01848-3905-4a61-8e2c-a192505e1b00",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SFigure 2: plot cross-validation\n",
    "\n",
    "better_auc$Model <- paste0(\"T_Inflamed+\", rownames(better_auc))\n",
    "\n",
    "p1 <- ggplot(data = better_auc, aes(x = Model, y = T_Inflamed)) + geom_bar(stat = \"identity\") +\n",
    "    geom_hline(yintercept = qbinom(p = 0.05, size = 100, prob = 0.5, lower.tail = F), color = \"blue\", linetype=\"dashed\") +\n",
    "    labs(title = \"Compare with T_Inflamed alone\", y = \"Frequency of weighted sum with higher AUC\", x = \"Model for computing weighted sum\") + coord_flip() + theme_bw()\n",
    "p2 <- ggplot(data = better_auc, aes(x = Model, y = NegSignature)) + geom_bar(stat = \"identity\") + \n",
    "    geom_hline(yintercept = qbinom(p = 0.05, size = 100, prob = 0.5, lower.tail = F), color = \"blue\", linetype=\"dashed\") +\n",
    "    labs(title = \"Compare with negative signature alone\", y = \"Frequency of weighted sum with higher AUC\", x = \"Model for computing weighted sum\") + coord_flip() + theme_bw() +\n",
    "    theme(axis.text.y = element_blank(), axis.title.y = element_blank())\n",
    "ggsave(\"Figure/SFigure2.pdf\", p1+p2, width = 10, height = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caa51e8d-77ef-4875-8b9e-f50959498b72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fc537b62-4b4a-4584-8c0d-88ff6f0970d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "========================================\n",
      "ComplexHeatmap version 2.18.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite either one:\n",
      "- Gu, Z. Complex Heatmap Visualization. iMeta 2022.\n",
      "- Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "    genomic data. Bioinformatics 2016.\n",
      "\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "! pheatmap() has been masked by ComplexHeatmap::pheatmap(). Most of the arguments\n",
      "   in the original pheatmap() are identically supported in the new function. You \n",
      "   can still use the original function by explicitly calling pheatmap::pheatmap().\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘ComplexHeatmap’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:pheatmap’:\n",
      "\n",
      "    pheatmap\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Figure 3\n",
    "\n",
    "library(ComplexHeatmap)\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "temp_rank <- data.frame(matrix(NA, ncol=length(study[[pname]]$pair_neg$gs), nrow=nrow(study[[pname]]$pheno)))\n",
    "colnames(temp_rank) <- paste0(\"T+\",study[[pname]]$pair_neg$gs)\n",
    "rownames(temp_rank) <- rownames(study[[pname]]$pheno)\n",
    "\n",
    "for(one_gs in study[[pname]]$pair_neg$gs){\n",
    "    ind_gs <- which(study[[pname]]$pair_neg$stat$v1 == one_gs)\n",
    "    pred_comb <- study[[pname]]$pair_neg$stat[ind_gs, \"b1\"]*study[[pname]]$pheno[[one_gs]]+\n",
    "        study[[pname]]$pair_neg$stat[ind_gs, \"b2\"]*study[[pname]]$pheno$T_Inflamed\n",
    "    temp_rank[[paste0(\"T+\",one_gs)]] <- rank(-pred_comb)-rank(-study[[pname]]$pheno$T_Inflamed)\n",
    "}\n",
    "\n",
    "temp_rank <- t(temp_rank)\n",
    "\n",
    "col.anno <- data.frame(\n",
    "    T_Inflamed_rank = rank(study[[pname]]$pheno$T_Inflamed), # rank from small to large to match the gradient coloring from white to dark red\n",
    "    Response = study[[pname]]$pheno$Response_Bin,\n",
    "    Immune_class = study[[pname]]$pheno$M_cluster\n",
    ")\n",
    "rownames(col.anno) <- colnames(temp_rank)\n",
    "\n",
    "\n",
    "set.seed(138)\n",
    "\n",
    "p1w <- ComplexHeatmap::pheatmap(temp_rank, color = colorRampPalette(RColorBrewer::brewer.pal(n = 7, name = \"PRGn\"))(100), show_colnames = F,\n",
    "                               annotation_legend = T, annotation_col = col.anno,\n",
    "                               border_color = \"grey60\", name=\"Rank_Diff\", row_names_max_width = unit(13, \"cm\"),\n",
    "                               fontsize = 8, treeheight_row = 5, treeheight_col = 5,\n",
    "                              heatmap_legend_param = list(legend_direction = \"horizontal\", legend_width = unit(3, \"cm\"), legend_gp = gpar(fontsize = 8)))\n",
    "\n",
    "plot_gs <- c(\"Salcher2022_non_classical_monocyte\",\"Salcher2022_neutrophil\",\"Bagaev2021_Matrix_remodeling\",\"Danaher2017_Neutrophil\",\n",
    "             \"T_Inflamed\")\n",
    "\n",
    "exp_mtx <- t(study[[pname]]$pheno[,plot_gs])\n",
    "exp_mtx[exp_mtx>1] <- 1\n",
    "exp_mtx[exp_mtx<-1] <- -1\n",
    "\n",
    "p2 <- ComplexHeatmap::pheatmap(exp_mtx, breaks = seq(-1,1,2/100), annotation_legend=F,\n",
    "                               color = colorRampPalette(rev(RColorBrewer::brewer.pal(n = 7, name = \"RdBu\")))(100), \n",
    "                               show_colnames = F, , border_color = \"grey60\", name=\"Exp_Z\", row_names_max_width = unit(13, \"cm\"),\n",
    "                               fontsize = 8, treeheight_row = 5,\n",
    "                               heatmap_legend_param = list(legend_direction = \"horizontal\", legend_width = unit(3, \"cm\"), legend_gp = gpar(fontsize = 8)))\n",
    "\n",
    "\n",
    "pdf(file=\"Figure/Figure3.pdf\", width = 12, height=7)\n",
    "ComplexHeatmap::draw(p1w %v% p2, heatmap_legend_side=\"bottom\", annotation_legend_side=\"bottom\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca97a942-33ad-46e3-b76b-d57e14ada987",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "fd890920-355a-49e8-b364-c57063244125",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in recalculate_width_panels(panel_positions, mapped_text = mapped_text, :\n",
      "“Unable to resize forest panel to be smaller than its heading; consider a smaller text size”\n",
      "Warning message in recalculate_width_panels(panel_positions, mapped_text = mapped_text, :\n",
      "“Unable to resize forest panel to be smaller than its heading; consider a smaller text size”\n",
      "Warning message in recalculate_width_panels(panel_positions, mapped_text = mapped_text, :\n",
      "“Unable to resize forest panel to be smaller than its heading; consider a smaller text size”\n",
      "Warning message in recalculate_width_panels(panel_positions, mapped_text = mapped_text, :\n",
      "“Unable to resize forest panel to be smaller than its heading; consider a smaller text size”\n"
     ]
    }
   ],
   "source": [
    "# Figure 4 and SFigure 1: survival plots (old)\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "for(one_gs in study[[pname]]$pair_neg$gs){\n",
    "    ind_gs <- which(study[[pname]]$pair_neg$stat$v1 == one_gs)\n",
    "    study[[pname]]$pheno[[paste0(\"T_\", one_gs)]] <- study[[pname]]$pair_neg$stat[ind_gs, \"b1\"]*study[[pname]]$pheno[[one_gs]]+\n",
    "        study[[pname]]$pair_neg$stat[ind_gs, \"b2\"]*study[[pname]]$pheno$T_Inflamed\n",
    "}\n",
    "\n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", paste0(\"T_\",study[[pname]]$pair_neg$gs)),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_OS_Months, Harmonized_OS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "p <- forest_model(model_list = univ_models,  merge_models =T)\n",
    "ggsave(\"Figure/Figure4A_old.pdf\", p, width=10, height=8)\n",
    "\n",
    "\n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", study[[pname]]$pair_neg$gs),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_OS_Months, Harmonized_OS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "p <- forest_model(model_list = univ_models,  merge_models =T)\n",
    "\n",
    "ggsave(\"Figure/SFigure1A_old.pdf\", p, width=10, height=8)\n",
    "\n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", paste0(\"T_\",study[[pname]]$pair_neg$gs)),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_PFS_Months, Harmonized_PFS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "p <- forest_model(model_list = univ_models,  merge_models =T)\n",
    "ggsave(\"Figure/Figure4B_old.pdf\", p, width=10, height=8)\n",
    "\n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", study[[pname]]$pair_neg$gs),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_PFS_Months, Harmonized_PFS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "p <- forest_model(model_list = univ_models,  merge_models =T)\n",
    "\n",
    "ggsave(\"Figure/SFigure1B_old.pdf\", p, width=10, height=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e378ae0-e691-497c-92a7-20ff7ebf78ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "59dc8196-c45b-47e7-8a87-6967ced10295",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions for forest plot\n",
    "\n",
    "get_scale <- function(plot, width_wanted, height_wanted, unit = \"in\"){\n",
    "  h <- convertHeight(sum(plot$heights), unit, TRUE)\n",
    "  w <- convertWidth(sum(plot$widths), unit, TRUE)\n",
    "  max(c(w/width_wanted,  h/height_wanted))\n",
    "}\n",
    "\n",
    "forest_plot <- function(model_list, model_names, xlim, ticks_at){\n",
    "    dt <- data.frame(\n",
    "        Model = model_names,\n",
    "        est = sapply(model_names, function(x){summary(model_list[[x]])$conf.int[1,1]}),\n",
    "        low = sapply(model_names, function(x){summary(model_list[[x]])$conf.int[1,3]}),\n",
    "        hi = sapply(model_names, function(x){summary(model_list[[x]])$conf.int[1,4]}),\n",
    "        P = as.character(round(sapply(model_names, function(x){summary(model_list[[x]])$coefficients[1,5]}),3)),\n",
    "        HR = paste(rep(\" \", 10), collapse = \" \"),\n",
    "        C_index = as.character(round(sapply(model_names, function(x){model_list[[x]]$concordance[\"concordance\"]}),3))\n",
    "    )\n",
    "\n",
    "    dt$`HR (95% CI)` <- paste0(round(dt$est,2),\" (\",round(dt$low,2),\" - \",round(dt$hi,2),\")\")\n",
    "    dt$P <- sapply(dt$P, function(x){if(x == \"0\") \"<0.001\" else x})\n",
    "\n",
    "    p <- forest(\n",
    "        dt[,c(\"Model\",\"HR\",\"HR (95% CI)\",\"P\",\"C_index\")],\n",
    "        est = list(dt$est), lower = list(dt$low), upper = list(dt$hi),\n",
    "        ci_column = 2, xlim = xlim, ticks_at = ticks_at, ref_line = 1,\n",
    "        arrow_lab = c(\"Better\", \"Worse\")\n",
    "    )\n",
    "    return(p)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2c9811d8-4d50-4a70-a897-e1df0e8601c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figure 4 and SFigure 4: survival plots\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "for(one_gs in study[[pname]]$pair_neg$gs){\n",
    "    ind_gs <- which(study[[pname]]$pair_neg$stat$v1 == one_gs)\n",
    "    study[[pname]]$pheno[[paste0(\"T_\", one_gs)]] <- study[[pname]]$pair_neg$stat[ind_gs, \"b1\"]*study[[pname]]$pheno[[one_gs]]+\n",
    "        study[[pname]]$pair_neg$stat[ind_gs, \"b2\"]*study[[pname]]$pheno$T_Inflamed\n",
    "}\n",
    "\n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", paste0(\"T_\",study[[pname]]$pair_neg$gs)),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_OS_Months, Harmonized_OS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "\n",
    "p <- forest_plot(model_list = univ_models, model_names = c(\"T_Inflamed\", paste0(\"T_\",study[[pname]]$pair_neg$gs)), xlim = c(0, 1.5), ticks_at = c(0.2, 1, 1.5))\n",
    "                   \n",
    "p_sc <- get_scale(plot = p, width_wanted = 10, height_wanted = 8, unit = \"in\")\n",
    "ggsave(filename = \"Figure/Figure4A.pdf\", plot = p, width = 10, height = 8, scale = p_sc)\n",
    "                   \n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", study[[pname]]$pair_neg$gs),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_OS_Months, Harmonized_OS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "p <- forest_plot(model_list = univ_models, model_names = c(\"T_Inflamed\", study[[pname]]$pair_neg$gs), xlim = c(0, 2), ticks_at = c(0.5, 1, 1.5, 2))\n",
    "\n",
    "p_sc <- get_scale(plot = p, width_wanted = 10, height_wanted = 8, unit = \"in\")\n",
    "ggsave(\"Figure/SFigure4A.pdf\", p, width=10, height=8, scale = p_sc)\n",
    "\n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", paste0(\"T_\",study[[pname]]$pair_neg$gs)),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_PFS_Months, Harmonized_PFS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "p <- forest_plot(model_list = univ_models, model_names = c(\"T_Inflamed\", paste0(\"T_\",study[[pname]]$pair_neg$gs)), xlim = c(0, 1.5), ticks_at = c(0.2, 1, 1.5))\n",
    "p_sc <- get_scale(plot = p, width_wanted = 10, height_wanted = 8, unit = \"in\")\n",
    "ggsave(\"Figure/Figure4B.pdf\", p, width=10, height=8, scale = p_sc)\n",
    "\n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", study[[pname]]$pair_neg$gs),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_PFS_Months, Harmonized_PFS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "p <- forest_plot(model_list = univ_models, model_names = c(\"T_Inflamed\", study[[pname]]$pair_neg$gs), xlim = c(0, 2), ticks_at = c(0.5, 1, 1.5, 2))\n",
    "p_sc <- get_scale(plot = p, width_wanted = 10, height_wanted = 8, unit = \"in\")\n",
    "ggsave(\"Figure/SFigure4B.pdf\", p, width=10, height=8, scale = p_sc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e12b2d2-025a-44a5-9cb9-f3abd231a81f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "3a642805-d54e-4ad2-9d2a-e47c262e58f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate Stabl input data for feature selection\n",
    "\n",
    "write.csv(data.frame(sid=rownames(study[[pname]]$pheno), Response_Bin=study[[pname]]$pheno$Response_Bin), \"Data/Stabl/response.csv\", row.names=F)\n",
    "\n",
    "neg_genes <- intersect(unique(unlist(icb_gsdb[study[[pname]]$pair_neg$gs])), rownames(study[[pname]]$exp))\n",
    "\n",
    "stabl_in <- cbind(\n",
    "    sid=rownames(study[[pname]]$pheno),\n",
    "    T_Inflamed=study[[pname]]$pheno$T_Inflamed,\n",
    "    t(study[[pname]]$exp[neg_genes, rownames(study[[pname]]$pheno)])\n",
    ")\n",
    "write.csv(stabl_in, \"Data/Stabl/T_Inflamed_paired_neg_genes.csv\", row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8138479f-ac38-45b3-afb4-bbca9e0fc115",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Construct immune-altered signatures\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "study[[pname]]$stabl <- list()\n",
    "\n",
    "# Read Stabl output\n",
    "\n",
    "stabl_out <- cbind(read.csv(\"Data/Stabl/Max STABL scores.csv\", row.names=1), beta=0, pval=1)\n",
    "stabl_out <- stabl_out[-which(rownames(stabl_out) == \"T_Inflamed\"),]\n",
    "\n",
    "temp_pheno <- cbind(study[[pname]]$pheno[,c(\"Response_Bin\",\"T_Inflamed\")], scale(t(study[[pname]]$exp[rownames(stabl_out),rownames(study[[pname]]$pheno)])))\n",
    "\n",
    "for(i in 1:nrow(stabl_out)){\n",
    "    one_gene <- rownames(stabl_out)[i]\n",
    "    if(grepl(\"\\\\-\", rownames(stabl_out)[i])){\n",
    "        one_gene <- gsub(\"\\\\-\",\"_\",rownames(stabl_out)[i])\n",
    "        colnames(temp_pheno)[i+2] <- one_gene\n",
    "    }\n",
    "    test <- summary(glm(as.formula(paste0(\"Response_Bin ~ \",one_gene)), family = \"binomial\", data = temp_pheno))\n",
    "    if(one_gene %in% rownames(test$coefficients)){\n",
    "        test_ind <- which(one_gene == rownames(test$coefficients))\n",
    "        stabl_out[i, \"beta\"] <- test$coefficients[test_ind,1]\n",
    "        stabl_out[i, \"pval\"] <- test$coefficients[test_ind,4]\n",
    "    }\n",
    "}\n",
    "\n",
    "# select Stabl-ranked features that were negatively associated with response\n",
    "\n",
    "study[[pname]]$stabl$stabl_out <- stabl_out\n",
    "study[[pname]]$stabl$neg_stat <- stabl_out[stabl_out$beta < 0,]\n",
    "\n",
    "# Predict response by 5 immune-altered signatures alone or combined with T_Inflamed\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "topn_uni <- data.frame(matrix(NA, ncol=4, nrow=length(topn_range)))\n",
    "rownames(topn_uni) <- paste0(\"Stabl_top\",topn_range)\n",
    "colnames(topn_uni) <- c(\"beta\",\"p\",\"AIC\",\"AUC\")\n",
    "topn_uni$beta <- 0\n",
    "topn_uni$p <- 1\n",
    "\n",
    "topn_pair <- data.frame(matrix(NA, ncol=6, nrow=length(topn_range)))\n",
    "rownames(topn_pair) <- paste0(\"Stabl_top\",topn_range)\n",
    "colnames(topn_pair) <- c(\"b1\",\"b2\",\"p1\",\"p2\",\"AIC\",\"AUC\")\n",
    "topn_pair$b1 <- 0\n",
    "topn_pair$b2 <- 0\n",
    "topn_pair$p1 <- 1\n",
    "topn_pair$p2 <- 1\n",
    "\n",
    "for(topn in topn_range){\n",
    "    study[[pname]]$stabl[[paste0(\"Stabl_top\",topn)]] <- rownames(study[[pname]]$stabl$neg_stat)[1:topn]\n",
    "    study[[pname]]$pheno[[paste0(\"Stabl_top\",topn)]] <- rowMeans(scale(t(study[[pname]]$exp[study[[pname]]$stabl[[paste0(\"Stabl_top\",topn)]],])))\n",
    "\n",
    "    test <- summary(glm(as.formula(paste0(\"Response_Bin ~ Stabl_top\",topn)), family = \"binomial\", data = study[[pname]]$pheno))\n",
    "    if(paste0(\"Stabl_top\",topn) %in% rownames(test$coefficients)){\n",
    "        test_ind <- which(paste0(\"Stabl_top\",topn) == rownames(test$coefficients))\n",
    "        topn_uni[paste0(\"Stabl_top\",topn), \"beta\"] <- test$coefficients[test_ind,1]\n",
    "        topn_uni[paste0(\"Stabl_top\",topn), \"p\"] <- test$coefficients[test_ind,4]\n",
    "    }\n",
    "    topn_uni[paste0(\"Stabl_top\",topn), \"AIC\"] <- test$aic\n",
    "    topn_uni[paste0(\"Stabl_top\",topn), \"AUC\"] <- auc(study[[pname]]$pheno$Response_Bin, -study[[pname]]$pheno[[paste0(\"Stabl_top\",topn)]], direction=\"<\")\n",
    "\n",
    "    test <- summary(glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed + Stabl_top\",topn)), family = \"binomial\", data = study[[pname]]$pheno))\n",
    "    if(paste0(\"T_Inflamed\") %in% rownames(test$coefficients)){\n",
    "        test_ind <- which(paste0(\"T_Inflamed\") == rownames(test$coefficients))\n",
    "        topn_pair[paste0(\"Stabl_top\",topn), \"b1\"] <- test$coefficients[test_ind,1]\n",
    "        topn_pair[paste0(\"Stabl_top\",topn), \"p1\"] <- test$coefficients[test_ind,4]\n",
    "    }\n",
    "    if(paste0(\"Stabl_top\",topn) %in% rownames(test$coefficients)){\n",
    "        test_ind <- which(paste0(\"Stabl_top\",topn) == rownames(test$coefficients))\n",
    "        topn_pair[paste0(\"Stabl_top\",topn), \"b2\"] <- test$coefficients[test_ind,1]\n",
    "        topn_pair[paste0(\"Stabl_top\",topn), \"p2\"] <- test$coefficients[test_ind,4]\n",
    "    }\n",
    "    topn_pair[paste0(\"Stabl_top\",topn), \"AIC\"] <- test$aic\n",
    "    study[[pname]]$pheno[[paste0(\"T_Inflamed_Stabl_top\",topn)]] <- study[[pname]]$pheno$T_Inflamed*topn_pair[paste0(\"Stabl_top\",topn), \"b1\"] +\n",
    "        study[[pname]]$pheno[[paste0(\"Stabl_top\",topn)]]*topn_pair[paste0(\"Stabl_top\",topn), \"b2\"]\n",
    "\n",
    "    topn_pair[paste0(\"Stabl_top\",topn), \"AUC\"] <- auc(study[[pname]]$pheno$Response_Bin, study[[pname]]$pheno[[paste0(\"T_Inflamed_Stabl_top\",topn)]], direction=\"<\")\n",
    "}\n",
    "\n",
    "study[[pname]]$stabl$topn_uni <- topn_uni\n",
    "study[[pname]]$stabl$topn_pair <- topn_pair"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1df0f89a-fc8a-4e32-aa0c-a0a6e5403106",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“The following requested variables were not found: SNORA81, NCF1B”\n"
     ]
    }
   ],
   "source": [
    "# Figure 5A: expression of Stabl-selected genes in NSCLC cell types\n",
    "\n",
    "selected_genes <- study$SU2C$stabl[[paste0(\"Stabl_top30\")]]\n",
    "Idents(sobj) <- \"cell_type\"\n",
    "p <- DotPlot(sobj, features = selected_genes) + RotatedAxis() + theme_bw() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))\n",
    "ggsave(\"Figure/Figure5A.pdf\", p, width=9, height=7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "96514629-1b46-4df5-8889-a9f3bd97fcd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Figures 5B and 5C: AIC and AUC for immune-altered signatures alone and when combining with T_Inflamed\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "dat.plot <- data.frame(\n",
    "    Model = c(rep(\"Univariate\", length(topn_range)+1), rep(\"Combined with\\nT_Inflamed\", length(topn_range))),\n",
    "    AIC = c(study[[pname]]$uni_gs[\"T_Inflamed\",\"AIC\"], study[[pname]]$stabl$topn_uni$AIC, study[[pname]]$stabl$topn_pair$AIC),\n",
    "    AUC = c(study[[pname]]$uni_gs[\"T_Inflamed\",\"AUC\"], study[[pname]]$stabl$topn_uni$AUC, study[[pname]]$stabl$topn_pair$AUC),\n",
    "    TopGene = c(\"T_Inflamed\", paste0(\"Top\",topn_range), paste0(\"Top\",topn_range))\n",
    ")\n",
    "dat.plot$Model <- factor(dat.plot$Model, levels = c(\"Univariate\",\"Combined with\\nT_Inflamed\"))\n",
    "\n",
    "p <- ggplot(data = dat.plot, aes(x=Model, y=AIC, group=TopGene, color=TopGene)) + geom_point() + geom_line() + theme_bw()\n",
    "ggsave(\"Figure/Figure5B.pdf\", p, width=4, height=3)\n",
    "\n",
    "p <- ggplot(data = dat.plot, aes(x=Model, y=AUC, group=TopGene, color=TopGene)) + geom_point() + geom_line() +\n",
    "    geom_hline(yintercept = sort(study[[pname]]$rand_model$auc_randg)[95], linetype=\"dashed\", color=\"darkgrey\") + theme_bw()\n",
    "ggsave(\"Figure/Figure5C.pdf\", p, width=4, height=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a55a1a8d-8350-46e5-860a-457c66ec5527",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SFigure 5\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "dat.plot <- data.frame(matrix(0, nrow = length(topn_range), ncol=2))\n",
    "rownames(dat.plot) <- paste0(\"Stabl_top\",topn_range)\n",
    "colnames(dat.plot) <- c(\"T_Inflamed\",\"Immune_altered\")\n",
    "\n",
    "lr.t <- glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed\")), family = \"binomial\", data = study[[pname]]$pheno)\n",
    "for(one_gs in paste0(\"Stabl_top\",topn_range)){\n",
    "    lr.uni <- glm(as.formula(paste0(\"Response_Bin ~ \", one_gs)), family = \"binomial\", data = study[[pname]]$pheno)\n",
    "    lr.comb <- glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed + \", one_gs)), family = \"binomial\", data = study[[pname]]$pheno)\n",
    "    dat.plot[one_gs, \"T_Inflamed\"] <- anova(lr.t,lr.comb,test=\"LRT\")[2,\"Pr(>Chi)\"]\n",
    "    dat.plot[one_gs, \"Immune_altered\"] <- anova(lr.uni,lr.comb,test=\"LRT\")[2,\"Pr(>Chi)\"]\n",
    "}\n",
    "\n",
    "dat.plot <- -log10(dat.plot)\n",
    "dat.plot$Model <- paste0(\"Response ~ T_Inflamed + \", paste0(\"Stabl_top\",topn_range))\n",
    "\n",
    "\n",
    "p1 <- ggplot(data = dat.plot, aes(x = Model, y = T_Inflamed)) + geom_bar(stat = \"identity\") + geom_hline(yintercept = -log10(0.05), color = \"blue\", linetype=\"dashed\") +\n",
    "    labs(title = \"Response ~ T_Inflamed\", y = \"-Log10P (LRT test)\") + coord_flip() + theme_bw()\n",
    "p2 <- ggplot(data = dat.plot, aes(x = Model, y = Immune_altered)) + geom_bar(stat = \"identity\") + geom_hline(yintercept = -log10(0.05), color = \"blue\", linetype=\"dashed\") +\n",
    "    labs(title = \"Response ~ immune-altered signature\", y = \"-Log10P (LRT test)\") + coord_flip() + theme_bw() + theme(axis.text.y = element_blank(), axis.title.y = element_blank())\n",
    "ggsave(\"Figure/SFigure5.pdf\", p1+p2, width = 10, height = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4f200fb9-3269-4f1c-90f9-85e0b06a143a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SFigure 6: predict OS and PFS using immune-altered signatures alone or combining with T_Inflamed\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", paste0(\"Stabl_top\",topn_range), paste0(\"T_Inflamed_Stabl_top\",topn_range)),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_OS_Months, Harmonized_OS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "p <- forest_plot(model_list = univ_models, model_names = c(\"T_Inflamed\", paste0(\"Stabl_top\",topn_range), paste0(\"T_Inflamed_Stabl_top\",topn_range)),\n",
    "                 xlim = c(0, 3.5), ticks_at = c(0.2, 1, 1.5, 2, 2.5, 3.5))\n",
    "                   \n",
    "p_sc <- get_scale(plot = p, width_wanted = 10, height_wanted = 8, unit = \"in\")\n",
    "ggsave(filename = \"Figure/SFigure6A.pdf\", plot = p, width = 10, height = 8, scale = p_sc)\n",
    "\n",
    "# p <- forest_model(model_list = univ_models,  merge_models =T)\n",
    "# ggsave(\"Figure/SFigure6A_old.pdf\", p, width=8, height=8)\n",
    "\n",
    "univ_formulas <- sapply(c(\"T_Inflamed\", paste0(\"Stabl_top\",topn_range), paste0(\"T_Inflamed_Stabl_top\",topn_range)),\n",
    "                        function(x){as.formula(paste('Surv(Harmonized_PFS_Months, Harmonized_PFS_Event)~', x))})\n",
    "univ_models <- lapply(univ_formulas, function(x){coxph(x, data = study[[pname]]$pheno)})\n",
    "\n",
    "p <- forest_plot(model_list = univ_models, model_names = c(\"T_Inflamed\", paste0(\"Stabl_top\",topn_range), paste0(\"T_Inflamed_Stabl_top\",topn_range)),\n",
    "                 xlim = c(0, 2.5), ticks_at = c(0.5, 1, 1.5, 2, 2.5))\n",
    "                   \n",
    "p_sc <- get_scale(plot = p, width_wanted = 10, height_wanted = 8, unit = \"in\")\n",
    "ggsave(filename = \"Figure/SFigure6B.pdf\", plot = p, width = 10, height = 8, scale = p_sc)\n",
    "\n",
    "# p <- forest_model(model_list = univ_models,  merge_models =T)\n",
    "# ggsave(\"Figure/SFigure6B_old.pdf\", p, width=8, height=8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e22c9cce-467d-4ac9-8527-d19e9b95e58a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "de89c0d7-6fd4-4299-9009-5b57e040e17f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Warning message:\n",
      "“glm.fit: fitted probabilities numerically 0 or 1 occurred”\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Warning message:\n",
      "“glm.fit: fitted probabilities numerically 0 or 1 occurred”\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n",
      "Setting levels: control = 0, case = 1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Compute combined models for the T cell inflamed GEP + immune-altered signatures\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "lrt_comp <- data.frame(matrix(0, nrow = length(topn_range), ncol=2))\n",
    "rownames(lrt_comp) <- paste0(\"Stabl_top\",topn_range)\n",
    "colnames(lrt_comp) <- c(\"T_Inflamed\",\"Immune_altered\")\n",
    "\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "\n",
    "tested_studies <- c(\"NSCLC_Advanced_GSE135222_Kim2020\",\"NSCLC_Metastatic_GSE136961_Hwang2020\",\n",
    "                    \"STAD_Metastatic_PRJEB25780_Kim2018\",\"Melanoma_Metastatic_PRJEB23709_Gide2019\",\"Melanoma_Advanced_GSE91061_Riaz2017\",\n",
    "                    \"Melanoma_Metastatic_GSE78220_Hugo2016\")\n",
    "\n",
    "study[[pname]]$stabl$val <- list()\n",
    "\n",
    "for(pname2 in tested_studies){\n",
    "    \n",
    "    study[[pname]]$stabl$val[[pname2]] <- list()\n",
    "\n",
    "    topn_uni <- data.frame(matrix(NA, ncol=4, nrow=length(topn_range)))\n",
    "    rownames(topn_uni) <- paste0(\"Stabl_top\",topn_range)\n",
    "    colnames(topn_uni) <- c(\"beta\",\"p\",\"AIC\",\"AUC\")\n",
    "    topn_uni$beta <- 0\n",
    "    topn_uni$p <- 1\n",
    "    \n",
    "    topn_val <- topn_uni\n",
    "\n",
    "    topn_pair <- data.frame(matrix(NA, ncol=6, nrow=length(topn_range)))\n",
    "    rownames(topn_pair) <- paste0(\"Stabl_top\",topn_range)\n",
    "    colnames(topn_pair) <- c(\"b1\",\"b2\",\"p1\",\"p2\",\"AIC\",\"AUC\")\n",
    "    topn_pair$b1 <- 0\n",
    "    topn_pair$b2 <- 0\n",
    "    topn_pair$p1 <- 1\n",
    "    topn_pair$p2 <- 1\n",
    "    \n",
    "    topn_lrt <- lrt_comp\n",
    "\n",
    "    for(topn in topn_range){\n",
    "        study[[pname2]]$pheno[[paste0(\"Stabl_top\",topn)]] <- 0\n",
    "        valid_genes <- intersect(study[[pname]]$stabl[[paste0(\"Stabl_top\",topn)]], rownames(study[[pname2]]$exp))\n",
    "        if(length(valid_genes) == 0){\n",
    "            next\n",
    "        }\n",
    "        \n",
    "        study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]] <- list()\n",
    "        study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]]$genes <- valid_genes\n",
    "        \n",
    "        temp_pheno <- data.frame(\n",
    "            Response_Bin = study[[pname2]]$pheno$Response_Bin,\n",
    "            T_Inflamed = study[[pname2]]$pheno$T_Inflamed,\n",
    "            Test = 0\n",
    "        )\n",
    "        \n",
    "        if(length(valid_genes) == 1){\n",
    "            temp_pheno$Test <- as.numeric(scale(as.numeric(study[[pname2]]$exp[valid_genes,])))\n",
    "        } else {\n",
    "            temp_pheno$Test <- rowMeans(scale(t(study[[pname2]]$exp[valid_genes,])))\n",
    "        }\n",
    "        study[[pname2]]$pheno[[paste0(\"Stabl_top\",topn)]] <- temp_pheno$Test\n",
    "        \n",
    "        lr.t <- glm(as.formula(paste0(\"Response_Bin ~ T_Inflamed\")), family = \"binomial\", data = temp_pheno)\n",
    "        \n",
    "        # combine T_Inflamed and immune-altered signatures using weights learned from SU2C\n",
    "        temp_pheno$CombinedVal <- temp_pheno$T_Inflamed*study[[pname]]$stabl$topn_pair[paste0(\"Stabl_top\",topn), \"b1\"] +\n",
    "            temp_pheno$Test*study[[pname]]$stabl$topn_pair[paste0(\"Stabl_top\",topn), \"b2\"]\n",
    "        \n",
    "        # predict response using immune-altered signature alone\n",
    "        lr.uni <- glm(as.formula(paste0(\"Response_Bin ~ Test\")), family = \"binomial\", data = temp_pheno)\n",
    "        test <- summary(lr.uni)\n",
    "        if(\"Test\" %in% rownames(test$coefficients)){\n",
    "            test_ind <- which(paste0(\"Test\") == rownames(test$coefficients))\n",
    "            topn_uni[paste0(\"Stabl_top\",topn), \"beta\"] <- test$coefficients[test_ind,1]\n",
    "            topn_uni[paste0(\"Stabl_top\",topn), \"p\"] <- test$coefficients[test_ind,4]\n",
    "        }\n",
    "        topn_uni[paste0(\"Stabl_top\",topn), \"AIC\"] <- test$aic\n",
    "        topn_uni[paste0(\"Stabl_top\",topn), \"AUC\"] <- auc(temp_pheno$Response_Bin, -temp_pheno$Test, direction=\"<\")\n",
    "        study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]]$uni <- test\n",
    "        \n",
    "        # predict response using T_Inflamed+immune-altered signature combined by SU2C weights\n",
    "        test <- summary(glm(as.formula(paste0(\"Response_Bin ~ CombinedVal\")), family = \"binomial\", data = temp_pheno))\n",
    "        if(\"CombinedVal\" %in% rownames(test$coefficients)){\n",
    "            test_ind <- which(paste0(\"CombinedVal\") == rownames(test$coefficients))\n",
    "            topn_val[paste0(\"Stabl_top\",topn), \"beta\"] <- test$coefficients[test_ind,1]\n",
    "            topn_val[paste0(\"Stabl_top\",topn), \"p\"] <- test$coefficients[test_ind,4]\n",
    "        }\n",
    "        topn_val[paste0(\"Stabl_top\",topn), \"AIC\"] <- test$aic\n",
    "        topn_val[paste0(\"Stabl_top\",topn), \"AUC\"] <- auc(temp_pheno$Response_Bin, temp_pheno$CombinedVal, direction=\"<\")\n",
    "        study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]]$val <- test\n",
    "\n",
    "        # predict response using T_Inflamed+immune-altered signature\n",
    "        lr.comb <- glm(as.formula(\"Response_Bin ~ T_Inflamed + Test\"), family = \"binomial\", data = temp_pheno)\n",
    "        test <- summary(lr.comb)\n",
    "        if(\"T_Inflamed\" %in% rownames(test$coefficients)){\n",
    "            test_ind <- which(\"T_Inflamed\" == rownames(test$coefficients))\n",
    "            topn_pair[paste0(\"Stabl_top\",topn), \"b1\"] <- test$coefficients[test_ind,1]\n",
    "            topn_pair[paste0(\"Stabl_top\",topn), \"p1\"] <- test$coefficients[test_ind,4]\n",
    "        }\n",
    "        if(\"Test\" %in% rownames(test$coefficients)){\n",
    "            test_ind <- which(\"Test\" == rownames(test$coefficients))\n",
    "            topn_pair[paste0(\"Stabl_top\",topn), \"b2\"] <- test$coefficients[test_ind,1]\n",
    "            topn_pair[paste0(\"Stabl_top\",topn), \"p2\"] <- test$coefficients[test_ind,4]\n",
    "        }\n",
    "        topn_pair[paste0(\"Stabl_top\",topn), \"AIC\"] <- test$aic\n",
    "        temp_pheno$Combined <- temp_pheno$T_Inflamed*topn_pair[paste0(\"Stabl_top\",topn), \"b1\"] +\n",
    "            temp_pheno$Test*topn_pair[paste0(\"Stabl_top\",topn), \"b2\"]\n",
    "\n",
    "        topn_pair[paste0(\"Stabl_top\",topn), \"AUC\"] <- auc(temp_pheno$Response_Bin, temp_pheno$Combined, direction=\"<\")\n",
    "        study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]]$pair <- test\n",
    "        \n",
    "        topn_lrt[paste0(\"Stabl_top\",topn), \"T_Inflamed\"] <- anova(lr.t,lr.comb,test=\"LRT\")[2,\"Pr(>Chi)\"]\n",
    "        topn_lrt[paste0(\"Stabl_top\",topn), \"Immune_altered\"] <- anova(lr.uni,lr.comb,test=\"LRT\")[2,\"Pr(>Chi)\"]\n",
    "    }\n",
    "\n",
    "    study[[pname]]$stabl$val[[pname2]]$topn_uni <- topn_uni\n",
    "    study[[pname]]$stabl$val[[pname2]]$topn_val <- topn_val\n",
    "    study[[pname]]$stabl$val[[pname2]]$topn_pair <- topn_pair\n",
    "    study[[pname]]$stabl$val[[pname2]]$topn_lrt <- topn_lrt\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c11493a1-4353-48de-87ce-084a99805a7f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "939ed447-e6d5-458c-a885-0fecb68602bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SFigure 7 and 8\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "\n",
    "# P-value changes\n",
    "\n",
    "for(pname2 in names(study[[pname]]$stabl$val)){\n",
    "    name_vec <- unlist(strsplit(pname2,\"_\"))\n",
    "    study_name <- paste0(name_vec[1],\"_\",name_vec[4])\n",
    "    \n",
    "    if(study_name == \"STAD_Kim2018\"){\n",
    "        study_name <- \"GC_Kim2018\"\n",
    "    }\n",
    "    \n",
    "    T_inf_label <- paste0(\"T_Inflamed (n=\",length(intersect(rownames(study[[pname2]]$exp), icb_gsdb$T_Inflamed)),\")\")\n",
    "    topn_label <- c()\n",
    "    for(topn in topn_range){\n",
    "        if(is.null(study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]])){\n",
    "            topn_label <- c(topn_label, paste0(\"Top\",topn,\" (n=0)\"))\n",
    "        } else {\n",
    "            topn_label <- c(topn_label, paste0(\"Top\",topn,\" (n=\",length(study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]]$genes),\")\"))\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    dat.plot <- data.frame(matrix(NA, ncol=5, nrow=2*length(topn_range)+1))\n",
    "    colnames(dat.plot) <- c(\"pval\",\"beta\",\"Type\",\"Signature\",\"Predictor\")\n",
    "    \n",
    "    dat.plot[1, \"pval\"] <- study[[pname2]]$uni_gs[\"T_Inflamed\",\"p\"]\n",
    "    dat.plot[1, \"beta\"] <- study[[pname2]]$uni_gs[\"T_Inflamed\",\"beta\"]\n",
    "    dat.plot[1, \"Signature\"] <- T_inf_label\n",
    "    dat.plot[1, \"Type\"] <- \"Signature alone\"\n",
    "    dat.plot[1, \"Predictor\"] <- \"T_Inflamed\"\n",
    "    \n",
    "    dat.plot[2:(1+length(topn_range)), \"pval\"] <- study[[pname]]$stabl$val[[pname2]]$topn_uni$p\n",
    "    dat.plot[2:(1+length(topn_range)), \"beta\"] <- study[[pname]]$stabl$val[[pname2]]$topn_uni$beta\n",
    "    dat.plot[2:(1+length(topn_range)), \"Signature\"] <- topn_label\n",
    "    dat.plot[2:(1+length(topn_range)), \"Type\"] <- \"Signature alone\"\n",
    "    dat.plot[2:(1+length(topn_range)), \"Predictor\"] <- paste0(\"Top\",topn_range)\n",
    "    \n",
    "    dat.plot[(1+length(topn_range)+1):(1+2*length(topn_range)), \"pval\"] <- study[[pname]]$stabl$val[[pname2]]$topn_val$p\n",
    "    dat.plot[(1+length(topn_range)+1):(1+2*length(topn_range)), \"beta\"] <- study[[pname]]$stabl$val[[pname2]]$topn_val$beta\n",
    "    dat.plot[(1+length(topn_range)+1):(1+2*length(topn_range)), \"Signature\"] <- topn_label\n",
    "    dat.plot[(1+length(topn_range)+1):(1+2*length(topn_range)), \"Type\"] <- \"Weighted difference\"\n",
    "    dat.plot[(1+length(topn_range)+1):(1+2*length(topn_range)), \"Predictor\"] <- paste0(\"T_Inf+Top\",topn_range,\"\\n(weighted_diff)\")\n",
    "    \n",
    "    dat.plot$Predictor <- factor(dat.plot$Predictor, levels=rev(c(\"T_Inflamed\",paste0(\"Top\",topn_range),paste0(\"T_Inf+Top\",topn_range,\"\\n(weighted_diff)\"))))\n",
    "    dat.plot$SignedLog10P <- -log10(dat.plot$pval)*sign(dat.plot$beta)\n",
    "    p <- ggplot(data = dat.plot, aes(x=Predictor, y=SignedLog10P, fill=Signature)) + geom_bar(stat=\"identity\") +\n",
    "        geom_hline(yintercept = c(-log10(0.05), 0, log10(0.05)), linetype=c(\"dashed\",\"solid\",\"dashed\"), color=c(\"blue\",\"darkgrey\",\"blue\")) +\n",
    "        scale_fill_manual(values = c(\"black\", RColorBrewer::brewer.pal(n=5,\"Set2\"))) + scale_color_manual(values = c()) +\n",
    "        labs(title=study_name) + theme_bw() + coord_flip()\n",
    "    ggsave(paste0(\"Figure/SFigure7_8_\",pname2,\".pdf\"), p, width=6, height=4)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "65ab84f8-4140-44ca-87f5-29eac2b45025",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SFigure 10 and 12\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "# P-value changes\n",
    "\n",
    "for(pname2 in names(study[[pname]]$stabl$val)){\n",
    "    name_vec <- unlist(strsplit(pname2,\"_\"))\n",
    "    study_name <- paste0(name_vec[1],\"_\",name_vec[4])\n",
    "    \n",
    "    if(study_name == \"STAD_Kim2018\"){\n",
    "        study_name <- \"GC_Kim2018\"\n",
    "    }\n",
    "    \n",
    "    T_inf_label <- paste0(\"T_Inflamed (n=\",length(intersect(rownames(study[[pname2]]$exp), icb_gsdb$T_Inflamed)),\")\")\n",
    "    topn_label <- c()\n",
    "    for(topn in topn_range){\n",
    "        if(is.null(study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]])){\n",
    "            topn_label <- c(topn_label, paste0(\"Top\",topn,\" (n=0)\"))\n",
    "        } else {\n",
    "            topn_label <- c(topn_label, paste0(\"Top\",topn,\" (n=\",length(study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]]$genes),\")\"))\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    dat.plot <- data.frame(matrix(NA, ncol=6, nrow=4*length(topn_range)))\n",
    "    colnames(dat.plot) <- c(\"pval\",\"beta\",\"Model\",\"Group\",\"Signature\",\"Type\")\n",
    "    \n",
    "    dat.plot[1:length(topn_range), \"pval\"] <- study[[pname2]]$uni_gs[\"T_Inflamed\",\"p\"]\n",
    "    dat.plot[1:length(topn_range), \"beta\"] <- study[[pname2]]$uni_gs[\"T_Inflamed\",\"beta\"]\n",
    "    dat.plot[1:length(topn_range), \"Model\"] <- \"Univariate\"\n",
    "    dat.plot[1:length(topn_range), \"Group\"] <- paste0(\"T_Inflamed:\",\"Top\",topn_range)\n",
    "    dat.plot[1:length(topn_range), \"Signature\"] <- T_inf_label\n",
    "    dat.plot[1:length(topn_range), \"Type\"] <- topn_label\n",
    "    \n",
    "    dat.plot[(1+length(topn_range)):(2*length(topn_range)), \"pval\"] <- study[[pname]]$stabl$val[[pname2]]$topn_uni$p\n",
    "    dat.plot[(1+length(topn_range)):(2*length(topn_range)), \"beta\"] <- study[[pname]]$stabl$val[[pname2]]$topn_uni$beta\n",
    "    dat.plot[(1+length(topn_range)):(2*length(topn_range)), \"Model\"] <- \"Univariate\"\n",
    "    dat.plot[(1+length(topn_range)):(2*length(topn_range)), \"Group\"] <- paste0(\"Top\",topn_range,\":\",\"Top\",topn_range)\n",
    "    dat.plot[(1+length(topn_range)):(2*length(topn_range)), \"Signature\"] <- \"Immune_altered\"\n",
    "    dat.plot[(1+length(topn_range)):(2*length(topn_range)), \"Type\"] <- topn_label\n",
    "    \n",
    "    dat.plot[(2*length(topn_range)+1):(length(topn_range)*3), \"pval\"] <- study[[pname]]$stabl$val[[pname2]]$topn_pair$p1\n",
    "    dat.plot[(2*length(topn_range)+1):(length(topn_range)*3), \"beta\"] <- study[[pname]]$stabl$val[[pname2]]$topn_pair$b1\n",
    "    dat.plot[(2*length(topn_range)+1):(length(topn_range)*3), \"Model\"] <- \"Combined\"\n",
    "    dat.plot[(2*length(topn_range)+1):(length(topn_range)*3), \"Group\"] <- paste0(\"T_Inflamed:\",\"Top\",topn_range)\n",
    "    dat.plot[(2*length(topn_range)+1):(length(topn_range)*3), \"Signature\"] <- T_inf_label\n",
    "    dat.plot[(2*length(topn_range)+1):(length(topn_range)*3), \"Type\"] <- topn_label\n",
    "    \n",
    "    dat.plot[(3*length(topn_range)+1):(length(topn_range)*4), \"pval\"] <- study[[pname]]$stabl$val[[pname2]]$topn_pair$p2\n",
    "    dat.plot[(3*length(topn_range)+1):(length(topn_range)*4), \"beta\"] <- study[[pname]]$stabl$val[[pname2]]$topn_pair$b2\n",
    "    dat.plot[(3*length(topn_range)+1):(length(topn_range)*4), \"Model\"] <- \"Combined\"\n",
    "    dat.plot[(3*length(topn_range)+1):(length(topn_range)*4), \"Group\"] <- paste0(\"Top\",topn_range,\":\",\"Top\",topn_range)\n",
    "    dat.plot[(3*length(topn_range)+1):(length(topn_range)*4), \"Signature\"] <- \"Immune_altered\"\n",
    "    dat.plot[(3*length(topn_range)+1):(length(topn_range)*4), \"Type\"] <- topn_label\n",
    "    \n",
    "    dat.plot$Model <- factor(dat.plot$Model, levels=c(\"Univariate\",\"Combined\"))\n",
    "    dat.plot$SignedLog10P <- -log10(dat.plot$pval)*sign(dat.plot$beta)\n",
    "    p <- ggplot(data = dat.plot, aes(x=Model, y=SignedLog10P, group=Group)) + geom_line(aes(color=Type)) +\n",
    "        geom_point(aes(shape=Signature), size=2) + labs(title=study_name) +\n",
    "        geom_hline(yintercept = c(-log10(0.05), 0, log10(0.05)), linetype=c(\"dashed\",\"solid\",\"dashed\"), color=c(\"blue\",\"darkgrey\",\"blue\")) +\n",
    "        theme_bw() + scale_color_manual(values = c(RColorBrewer::brewer.pal(n=5,\"Set2\")))\n",
    "    ggsave(paste0(\"Figure/SFigure10_12_\",pname2,\".pdf\"), p, width=6, height=4)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6c74186a-b5cc-47e4-8bf4-af0700df21c3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3 rows containing missing values (`geom_point()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3 rows containing missing values (`geom_line()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3 rows containing missing values (`geom_point()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 3 rows containing missing values (`geom_line()`).”\n"
     ]
    }
   ],
   "source": [
    "# Figure 6 and SFigure 9\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "for(pname2 in names(study[[pname]]$stabl$val)){\n",
    "    name_vec <- unlist(strsplit(pname2,\"_\"))\n",
    "    study_name <- paste0(name_vec[1],\"_\",name_vec[4])\n",
    "    \n",
    "    if(study_name == \"STAD_Kim2018\"){\n",
    "        study_name <- \"GC_Kim2018\"\n",
    "    }\n",
    "    \n",
    "    T_inf_label <- paste0(\"T_Inflamed (n=\",length(intersect(rownames(study[[pname2]]$exp), icb_gsdb$T_Inflamed)),\")\")\n",
    "    topn_label <- c()\n",
    "    for(topn in topn_range){\n",
    "        if(is.null(study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]])){\n",
    "            topn_label <- c(topn_label, paste0(\"Top\",topn,\" (n=0)\"))\n",
    "        } else {\n",
    "            topn_label <- c(topn_label, paste0(\"Top\",topn,\" (n=\",length(study[[pname]]$stabl$val[[pname2]][[paste0(\"Stabl_top\",topn)]]$genes),\")\"))\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    dat.plot <- data.frame(\n",
    "        Model = c(rep(\"Univariate\", length(topn_range)+1), rep(\"Combined\\n(Trained_Weight)\", length(topn_range)), rep(\"Univariate\\n(SU2C_Weight)\", length(topn_range))),\n",
    "        AIC = c(study[[pname2]]$uni_gs[\"T_Inflamed\",\"AIC\"], study[[pname]]$stabl$val[[pname2]]$topn_uni$AIC, study[[pname]]$stabl$val[[pname2]]$topn_pair$AIC, study[[pname]]$stabl$val[[pname2]]$topn_val$AIC),\n",
    "        AUC = c(study[[pname2]]$uni_gs[\"T_Inflamed\",\"AUC\"], study[[pname]]$stabl$val[[pname2]]$topn_uni$AUC, study[[pname]]$stabl$val[[pname2]]$topn_pair$AUC, study[[pname]]$stabl$val[[pname2]]$topn_val$AUC),\n",
    "        TopGene = c(T_inf_label, topn_label, topn_label, topn_label)\n",
    "    )\n",
    "    dat.plot$Model <- factor(dat.plot$Model, levels = c(\"Univariate\",\"Univariate\\n(SU2C_Weight)\",\"Combined\\n(Trained_Weight)\"))\n",
    "\n",
    "    p <- ggplot(data = dat.plot, aes(x=Model, y=AIC, group=TopGene, color=TopGene)) + geom_point() + geom_line() + theme_bw() +\n",
    "        labs(title=study_name) + scale_color_manual(values = c(\"black\", RColorBrewer::brewer.pal(n=5,\"Set2\"))) +\n",
    "        geom_hline(yintercept = study[[pname2]]$uni_gs[\"T_Inflamed\",\"AIC\"], linetype=\"dashed\", color=\"black\")\n",
    "    ggsave(paste0(\"Figure/Figure6ABC_SFigure9ABC_\",pname2,\".pdf\"), p, width=6, height=3)\n",
    "\n",
    "    p <- ggplot(data = dat.plot, aes(x=Model, y=AUC, group=TopGene, color=TopGene)) + geom_point() + geom_line() + labs(title=study_name) +\n",
    "        geom_hline(yintercept = sort(study[[pname2]]$rand_model$auc_randg)[95], linetype=\"dashed\", color=\"darkgrey\") + theme_bw() +\n",
    "        scale_color_manual(values = c(\"black\", RColorBrewer::brewer.pal(n=5,\"Set2\")))\n",
    "    ggsave(paste0(\"Figure/Figure6DEF_SFigure9DEF_\",pname2,\".pdf\"), p, width=6, height=3)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3f0118d-0e06-4314-baad-cfbc65c41d5d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "46ceafb5-f88c-4af2-a3bb-67d563825d9a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 rows containing missing values (`position_stack()`).”\n",
      "Warning message:\n",
      "“\u001b[1m\u001b[22mRemoved 1 rows containing missing values (`position_stack()`).”\n"
     ]
    }
   ],
   "source": [
    "# SFigure 11 and 13\n",
    "\n",
    "pname <- \"SU2C\"\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "for(pname2 in names(study[[pname]]$stabl$val)){\n",
    "    name_vec <- unlist(strsplit(pname2,\"_\"))\n",
    "    study_name <- paste0(name_vec[1],\"_\",name_vec[4])\n",
    "    \n",
    "    if(study_name == \"STAD_Kim2018\"){\n",
    "        study_name <- \"GC_Kim2018\"\n",
    "    }\n",
    "\n",
    "    dat.plot <- study[[pname]]$stabl$val[[pname2]]$topn_lrt\n",
    "    \n",
    "    if(any(dat.plot == 0)){\n",
    "        dat.plot[dat.plot == 0] <- NA\n",
    "    }\n",
    "    dat.plot <- -log10(dat.plot)\n",
    "    dat.plot$Model <- paste0(\"Response ~ T_Inflamed + \", paste0(\"Stabl_top\",topn_range))\n",
    "\n",
    "\n",
    "    p1 <- ggplot(data = dat.plot, aes(x = Model, y = T_Inflamed)) + geom_bar(stat = \"identity\") + geom_hline(yintercept = -log10(0.05), color = \"blue\", linetype=\"dashed\") +\n",
    "        labs(title = \"Response ~ T_Inflamed\", y = \"-Log10P (LRT test)\") + coord_flip() + theme_bw()\n",
    "    p2 <- ggplot(data = dat.plot, aes(x = Model, y = Immune_altered)) + geom_bar(stat = \"identity\") + geom_hline(yintercept = -log10(0.05), color = \"blue\", linetype=\"dashed\") +\n",
    "        labs(title = \"Response ~ Immune-altered signature\", y = \"-Log10P (LRT test)\") + coord_flip() + theme_bw() + theme(axis.text.y = element_blank(), axis.title.y = element_blank())\n",
    "    ggsave(paste0(\"Figure/SFigure11_13_\",study_name,\".pdf\"), p1+p2, width = 10, height = 3)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d641f5b-4a15-4843-b534-e2c6f4ca3e87",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "949a23b1-f4e2-4676-848a-623657f7d64e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65e6fcad-b919-4cfc-a5a8-c613e8387e0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e144fcf6-94a5-4106-ac5e-660b1f63b588",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Supplementary tables for model details\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "\n",
    "out_cols <- c(\"sid\",\"Response_Bin\",\"Harmonized_Confirmed_BOR_3_Cat\",\"Harmonized_OS_Months\", \"Harmonized_OS_Event\",\"Harmonized_PFS_Months\",\"Harmonized_PFS_Event\",\"M_cluster\",\n",
    "              \"Sex\",\"Smoking_Status\",\"Histology\",\"PDL1_TPS\",\"Age_at_Diagnosis\",\"Line_of_Therapy\",rownames(study$SU2C$uni_gs),paste0(\"Stabl_top\",topn_range))\n",
    "\n",
    "temp <- study$SU2C$pheno[, out_cols]\n",
    "write.csv(temp, \"Table/su2c_pheno_all_signature.csv\", row.names=F)\n",
    "\n",
    "temp <- study$SU2C$pair_neg$stat[,c(1:6,9:10)]\n",
    "write.csv(temp, \"Table/su2c_GEP_myeloid.csv\", row.names=F)\n",
    "temp <- rbind(study$SU2C$uni_gs[\"T_Inflamed\",],study$SU2C$uni_gs[study$SU2C$pair_neg$gs,])\n",
    "write.csv(temp, \"Table/su2c_myeloid_uni.csv\", row.names=T)\n",
    "\n",
    "temp <- cbind(data.frame(v1=\"T_Inflamed\",v2=c(rownames(study$SU2C$stabl$topn_pair))),study$SU2C$stabl$topn_pair)\n",
    "write.csv(temp, \"Table/su2c_GEP_stabl.csv\", row.names=F)\n",
    "temp <- rbind(study$SU2C$uni_gs[\"T_Inflamed\",], study$SU2C$stabl$topn_uni)\n",
    "write.csv(temp, \"Table/su2c_stabl_uni.csv\", row.names=T)\n",
    "\n",
    "pname <- \"NSCLC_Advanced_GSE135222_Kim2020\"\n",
    "temp <- cbind(data.frame(v1=\"T_Inflamed\",v2=c(rownames(study$SU2C$stabl$val[[pname]]$topn_pair))),study$SU2C$stabl$val[[pname]]$topn_pair)\n",
    "write.csv(temp, \"Table/Kim2020_GEP_stabl.csv\", row.names=F)\n",
    "temp <- rbind(study[[pname]]$uni_gs[\"T_Inflamed\",],study$SU2C$stabl$val[[pname]]$topn_uni)\n",
    "write.csv(temp, \"Table/Kim2020_uni.csv\", row.names=T)\n",
    "temp <- study$SU2C$stabl$val[[pname]]$topn_val\n",
    "write.csv(temp, \"Table/Kim2020_uni_su2c.csv\", row.names=T)\n",
    "\n",
    "pname <- \"NSCLC_Metastatic_GSE136961_Hwang2020\"\n",
    "temp <- cbind(data.frame(v1=\"T_Inflamed\",v2=c(rownames(study$SU2C$stabl$val[[pname]]$topn_pair))),study$SU2C$stabl$val[[pname]]$topn_pair)\n",
    "write.csv(temp, \"Table/Hwang2020_GEP_stabl.csv\", row.names=F)\n",
    "temp <- rbind(study[[pname]]$uni_gs[\"T_Inflamed\",],study$SU2C$stabl$val[[pname]]$topn_uni)\n",
    "write.csv(temp, \"Table/Hwang2020_uni.csv\", row.names=T)\n",
    "temp <- study$SU2C$stabl$val[[pname]]$topn_val\n",
    "write.csv(temp, \"Table/Hwang2020_uni_su2c.csv\", row.names=T)\n",
    "\n",
    "pname <- \"STAD_Metastatic_PRJEB25780_Kim2018\"\n",
    "temp <- cbind(data.frame(v1=\"T_Inflamed\",v2=c(rownames(study$SU2C$stabl$val[[pname]]$topn_pair))),study$SU2C$stabl$val[[pname]]$topn_pair)\n",
    "write.csv(temp, \"Table/Kim2018_GEP_stabl.csv\", row.names=F)\n",
    "temp <- rbind(study[[pname]]$uni_gs[\"T_Inflamed\",],study$SU2C$stabl$val[[pname]]$topn_uni)\n",
    "write.csv(temp, \"Table/Kim2018_uni.csv\", row.names=T)\n",
    "temp <- study$SU2C$stabl$val[[pname]]$topn_val\n",
    "write.csv(temp, \"Table/Kim2018_uni_su2c.csv\", row.names=T)\n",
    "\n",
    "pname <- \"Melanoma_Metastatic_PRJEB23709_Gide2019\"\n",
    "temp <- cbind(data.frame(v1=\"T_Inflamed\",v2=c(rownames(study$SU2C$stabl$val[[pname]]$topn_pair))),study$SU2C$stabl$val[[pname]]$topn_pair)\n",
    "write.csv(temp, \"Table/Gide2019_GEP_stabl.csv\", row.names=F)\n",
    "temp <- rbind(study[[pname]]$uni_gs[\"T_Inflamed\",],study$SU2C$stabl$val[[pname]]$topn_uni)\n",
    "write.csv(temp, \"Table/Gide2019_uni.csv\", row.names=T)\n",
    "temp <- study$SU2C$stabl$val[[pname]]$topn_val\n",
    "write.csv(temp, \"Table/Gide2019_uni_su2c.csv\", row.names=T)\n",
    "\n",
    "pname <- \"Melanoma_Advanced_GSE91061_Riaz2017\"\n",
    "temp <- cbind(data.frame(v1=\"T_Inflamed\",v2=c(rownames(study$SU2C$stabl$val[[pname]]$topn_pair))),study$SU2C$stabl$val[[pname]]$topn_pair)\n",
    "write.csv(temp, \"Table/Riaz2017_GEP_stabl.csv\", row.names=F)\n",
    "temp <- rbind(study[[pname]]$uni_gs[\"T_Inflamed\",],study$SU2C$stabl$val[[pname]]$topn_uni)\n",
    "write.csv(temp, \"Table/Riaz2017_uni.csv\", row.names=T)\n",
    "temp <- study$SU2C$stabl$val[[pname]]$topn_val\n",
    "write.csv(temp, \"Table/Riaz2017_uni_su2c.csv\", row.names=T)\n",
    "\n",
    "pname <- \"Melanoma_Metastatic_GSE78220_Hugo2016\"\n",
    "temp <- cbind(data.frame(v1=\"T_Inflamed\",v2=c(rownames(study$SU2C$stabl$val[[pname]]$topn_pair))),study$SU2C$stabl$val[[pname]]$topn_pair)\n",
    "write.csv(temp, \"Table/Hugo2016_GEP_stabl.csv\", row.names=F)\n",
    "temp <- rbind(study[[pname]]$uni_gs[\"T_Inflamed\",],study$SU2C$stabl$val[[pname]]$topn_uni)\n",
    "write.csv(temp, \"Table/Hugo2016_uni.csv\", row.names=T)\n",
    "temp <- study$SU2C$stabl$val[[pname]]$topn_val\n",
    "write.csv(temp, \"Table/Hugo2016_uni_su2c.csv\", row.names=T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "20fc1038-2655-481c-a866-93e074f48149",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Supplementary table to demonstrate improvements of combined models\n",
    "\n",
    "n_r <- sum(study$SU2C$pheno$Response_Bin == 1)\n",
    "n_nr <- sum(study$SU2C$pheno$Response_Bin == 0)\n",
    "med <- as.integer((nrow(study$SU2C$pheno)+1)/2)\n",
    "\n",
    "temp <- data.frame(matrix(NA, ncol=7, nrow=1+nrow(study$SU2C$pair_neg$stat)))\n",
    "colnames(temp) <- c(\"Signature\",\"TP\",\"TPR\",\"TN\",\"TNR\",\"ACC\",\"AUC\")\n",
    "\n",
    "temp[1, \"Signature\"] <- \"T_Inflamed\"\n",
    "temp[1, \"TP\"] <- sum(study$SU2C$pheno$Response_Bin[order(study$SU2C$pheno$T_Inflamed, decreasing = T)[1:med]] == 1)\n",
    "temp[1, \"TN\"] <- sum(study$SU2C$pheno$Response_Bin[order(study$SU2C$pheno$T_Inflamed, decreasing = T)[(med+1):(n_r+n_nr)]] == 0)\n",
    "temp[1, \"TPR\"] <- temp[1, \"TP\"]/med\n",
    "temp[1, \"TNR\"] <- temp[1, \"TN\"]/med\n",
    "temp[1, \"ACC\"] <- (temp[1, \"TP\"]+temp[1, \"TN\"])/(n_r+n_nr)\n",
    "temp[1, \"AUC\"] <- study$SU2C$uni_gs[\"T_Inflamed\", \"AUC\"]\n",
    "\n",
    "for(i in 1:nrow(study$SU2C$pair_neg$stat)){\n",
    "    temp_pred <- study$SU2C$pair_neg$stat[i, \"b1\"]*study$SU2C$pheno[[study$SU2C$pair_neg$stat[i, \"v1\"]]]+\n",
    "        study$SU2C$pair_neg$stat[i, \"b2\"]*study$SU2C$pheno[[study$SU2C$pair_neg$stat[i, \"v2\"]]]\n",
    "    \n",
    "    temp[1+i, \"Signature\"] <- paste0(study$SU2C$pair_neg$stat[i, \"v2\"],\"+\",study$SU2C$pair_neg$stat[i, \"v1\"])\n",
    "    temp[1+i, \"TP\"] <- sum(study$SU2C$pheno$Response_Bin[order(temp_pred, decreasing = T)[1:med]] == 1)\n",
    "    temp[1+i, \"TN\"] <- sum(study$SU2C$pheno$Response_Bin[order(temp_pred, decreasing = T)[(med+1):(n_r+n_nr)]] == 0)\n",
    "    temp[1+i, \"TPR\"] <- temp[1+i, \"TP\"]/med\n",
    "    temp[1+i, \"TNR\"] <- temp[1+i, \"TN\"]/med\n",
    "    temp[1+i, \"ACC\"] <- (temp[1+i, \"TP\"]+temp[1+i, \"TN\"])/(n_r+n_nr)\n",
    "    temp[1+i, \"AUC\"] <- study$SU2C$pair_neg$stat[i, \"AUC\"]\n",
    "}\n",
    "\n",
    "temp$TPR_inc <- temp$TPR - temp[1,\"TPR\"]\n",
    "temp$TNR_inc <- temp$TNR - temp[1,\"TNR\"]\n",
    "temp$ACC_inc <- temp$ACC - temp[1,\"ACC\"]\n",
    "temp$predicted <- med\n",
    "temp$total <- n_r+n_nr\n",
    "\n",
    "temp <- temp[,c(\"Signature\",\"TP\",\"predicted\",\"total\",\"TPR\",\"TPR_inc\")]\n",
    "colnames(temp) <- c(\"Model\",\"nTP\",\"nPos\",\"nTotal\",\"TPR (Response Rate)\",\"TPR Increase (vs. T_Inflamed)\")\n",
    "\n",
    "write.csv(temp, \"Table/improvement_myeloid_SU2C.csv\", row.names=F)\n",
    "\n",
    "\n",
    "topn_range <- c(10,15,20,25,30)\n",
    "for(pname2 in names(study$SU2C$stabl$val)){\n",
    "    n_r <- sum(study[[pname2]]$pheno$Response_Bin == 1)\n",
    "    n_nr <- sum(study[[pname2]]$pheno$Response_Bin == 0)\n",
    "    med <- as.integer((nrow(study[[pname2]]$pheno)+1)/2)\n",
    "\n",
    "    temp <- data.frame(matrix(NA, ncol=7, nrow=1+length(topn_range)))\n",
    "    colnames(temp) <- c(\"Signature\",\"TP\",\"TPR\",\"TN\",\"TNR\",\"ACC\",\"AUC\")\n",
    "    rownames(temp) <- c(\"T_Inflamed\", paste0(\"T_Stabl_top\",topn_range))\n",
    "\n",
    "    temp[\"T_Inflamed\", \"Signature\"] <- \"T_Inflamed\"\n",
    "    temp[\"T_Inflamed\", \"TP\"] <- sum(study[[pname2]]$pheno$Response_Bin[order(study[[pname2]]$pheno$T_Inflamed, decreasing = T)[1:med]] == 1)\n",
    "    temp[\"T_Inflamed\", \"TN\"] <- sum(study[[pname2]]$pheno$Response_Bin[order(study[[pname2]]$pheno$T_Inflamed, decreasing = T)[(med+1):(n_r+n_nr)]] == 0)\n",
    "    temp[\"T_Inflamed\", \"TPR\"] <- temp[1, \"TP\"]/med\n",
    "    temp[\"T_Inflamed\", \"TNR\"] <- temp[1, \"TN\"]/med\n",
    "    temp[\"T_Inflamed\", \"ACC\"] <- (temp[1, \"TP\"]+temp[1, \"TN\"])/(n_r+n_nr)\n",
    "    temp[\"T_Inflamed\", \"AUC\"] <- study[[pname2]]$uni_gs[\"T_Inflamed\", \"AUC\"]\n",
    "\n",
    "    for(topn in topn_range){\n",
    "        temp_pred <- study$SU2C$stabl$val[[pname2]]$topn_pair[paste0(\"Stabl_top\",topn), \"b1\"]*study[[pname2]]$pheno[[\"T_Inflamed\"]]+\n",
    "            study$SU2C$stabl$val[[pname2]]$topn_pair[paste0(\"Stabl_top\",topn), \"b2\"]*study[[pname2]]$pheno[[paste0(\"Stabl_top\",topn)]]\n",
    "        \n",
    "        if(all(temp_pred == 0)) temp_pred <- study[[pname2]]$pheno[[\"T_Inflamed\"]]\n",
    "\n",
    "        temp[paste0(\"T_Stabl_top\",topn), \"Signature\"] <- paste0(\"T_Inflamed + Stabl_top\",topn)\n",
    "        temp[paste0(\"T_Stabl_top\",topn), \"TP\"] <- sum(study[[pname2]]$pheno$Response_Bin[order(temp_pred, decreasing = T)[1:med]] == 1)\n",
    "        temp[paste0(\"T_Stabl_top\",topn), \"TN\"] <- sum(study[[pname2]]$pheno$Response_Bin[order(temp_pred, decreasing = T)[(med+1):(n_r+n_nr)]] == 0)\n",
    "        temp[paste0(\"T_Stabl_top\",topn), \"TPR\"] <- temp[paste0(\"T_Stabl_top\",topn), \"TP\"]/med\n",
    "        temp[paste0(\"T_Stabl_top\",topn), \"TNR\"] <- temp[paste0(\"T_Stabl_top\",topn), \"TN\"]/med\n",
    "        temp[paste0(\"T_Stabl_top\",topn), \"ACC\"] <- (temp[paste0(\"T_Stabl_top\",topn), \"TP\"]+temp[paste0(\"T_Stabl_top\",topn), \"TN\"])/(n_r+n_nr)\n",
    "        temp[paste0(\"T_Stabl_top\",topn), \"AUC\"] <- study$SU2C$stabl$val[[pname2]]$topn_uni[paste0(\"T_Stabl_top\",topn), \"AUC\"]\n",
    "    }\n",
    "\n",
    "    temp$TPR_inc <- temp$TPR - temp[\"T_Inflamed\",\"TPR\"]\n",
    "    temp$TNR_inc <- temp$TNR - temp[\"T_Inflamed\",\"TNR\"]\n",
    "    temp$ACC_inc <- temp$ACC - temp[\"T_Inflamed\",\"ACC\"]\n",
    "    temp$predicted <- med\n",
    "    temp$total <- n_r+n_nr\n",
    "\n",
    "    temp <- temp[,c(\"Signature\",\"TP\",\"predicted\",\"total\",\"TPR\",\"TPR_inc\")]\n",
    "    colnames(temp) <- c(\"Model\",\"nTP\",\"nPos\",\"nTotal\",\"TPR (Response Rate)\",\"TPR Increase (vs. T_Inflamed)\")\n",
    "\n",
    "    write.csv(temp, paste0(\"Table/improvement_Stabl_\",pname2,\".csv\"), row.names=F)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "2cd857a3-0878-4385-a750-a469f6276b32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# supplementary table for 170 signatures\n",
    "\n",
    "temp <- data.frame(Name = names(icb_gsdb), Category = stringr::str_split_fixed(names(icb_gsdb),\"_\",2)[,1], PMID = \"\", Reference = \"\")\n",
    "write.csv(temp, \"Table/signature_info.csv\", row.names = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e4b2777-b912-4da6-b7f1-7fdb64cc656d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0c75a753-568d-43a2-a613-572b60857c83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.3.2 (2023-10-31)\n",
       "Platform: x86_64-pc-linux-gnu (64-bit)\n",
       "Running under: Ubuntu 18.04.5 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1 \n",
       "LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1\n",
       "\n",
       "locale:\n",
       " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
       " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
       " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
       " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
       " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
       "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
       "\n",
       "time zone: Etc/UTC\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       "[1] stats     graphics  grDevices utils     datasets  methods   base     \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       " [1] digest_0.6.34   IRdisplay_1.1   utf8_1.2.4      base64enc_0.1-3\n",
       " [5] fastmap_1.1.1   glue_1.7.0      htmltools_0.5.7 repr_1.1.6     \n",
       " [9] lifecycle_1.0.4 cli_3.6.2       fansi_1.0.6     vctrs_0.6.5    \n",
       "[13] pbdZMQ_0.3-11   compiler_4.3.2  tools_4.3.2     evaluate_0.23  \n",
       "[17] pillar_1.9.0    crayon_1.5.2    rlang_1.1.3     jsonlite_1.8.8 \n",
       "[21] IRkernel_1.3.2  uuid_1.2-0     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8652ec7c-c4d2-42ac-b83f-9311029ff86e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
